{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6840c6bd",
   "metadata": {},
   "source": [
    "# DESeq2 with ~group as design factor, included whole dataset \n",
    "## Use contrast () function to make desired control vs timepoint comparisons \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ecd8df53",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Installing and loading the required packages at once\n",
    "\n",
    "if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "  install.packages(\"BiocManager\")\n",
    "\n",
    "#BiocManager::install(\"DESeq2\")\n",
    "require(DESeq2)\n",
    "#BiocManager::install(\"org.Hs.eg.db\")\n",
    "require(org.Hs.eg.db)\n",
    "#BiocManager::install(\"EnhancedVolcano\")\n",
    "require(EnhancedVolcano)\n",
    "#install.packages(\"RColorBrewer\")\n",
    "\n",
    "require(RColorBrewer)\n",
    "#install.packages(\"ggplot2\")\n",
    "require(ggplot2)\n",
    "#BiocManager::install(\"Glimma\")\n",
    "require(Glimma)\n",
    "#BiocManager::install(\"ComplexHeatmap\")\n",
    "require(ComplexHeatmap)\n",
    "#install.packages(\"gprofiler2\")\n",
    "library(gprofiler2)\n",
    "#BiocManager::install(\"tidyverse\")\n",
    "library(tidyverse)\n",
    "#BiocManager::install(\"tradeSeq\", lib= \"C:/Users/cb998/R-4.1.2/library\")\n",
    "require(tradeSeq)\n",
    "library(dplyr)\n",
    "library(data.table)\n",
    "library(tidyverse)\n",
    "require(pander)\n",
    "library(plyr)\n",
    "library(readr) \n",
    "library(AnnotationDbi)\n",
    "library(org.Hs.eg.db)\n",
    "library(EnhancedVolcano)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4a4c6fe3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Set working directory ####\n",
    "setwd('C:/Users/issyl/OneDrive - University of Cambridge/Research Project/DGE analysis AAI/DESeq2 analysis/RNA-Seq-AAI-analysis/00_raw data')\n",
    "\n",
    "## Loading counts matrix and metadata\n",
    "cnts <- read.delim(file.choose(), row.names = 1)\n",
    "coldata<- read.delim(file.choose(), row.names=1)\n",
    "\n",
    "### Use this for patients study samples\n",
    "coldata <- coldata[,c(\"treatment\", \"timepoint\", \"stage\", \"ID\",\"line\",\"group\")]\n",
    "\n",
    "# Make factors \n",
    "coldata$treatment <- factor(coldata$treatment)\n",
    "coldata$timepoint <- factor(coldata$timepoint)\n",
    "coldata$stage <- factor(coldata$stage)\n",
    "coldata$ID <- factor(coldata$ID)\n",
    "coldata$line <- factor(coldata$line)\n",
    "coldata$group <- factor(coldata$group)\n",
    "\n",
    "### Make a DESeq2 design \n",
    "dds <- DESeqDataSetFromMatrix(countData = cnts,\n",
    "                              colData = coldata,\n",
    "                              design = ~ group)\n",
    "\n",
    "#### Set factor level \n",
    "dds$group <- relevel(dds$group, ref=\"fetal.control\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "c4eaa34f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "log2 fold change (MLE): group paed.control vs fetal.control \n",
       "Wald test p-value: group paed.control vs fetal.control \n",
       "DataFrame with 6 rows and 6 columns\n",
       "                     baseMean log2FoldChange     lfcSE      stat    pvalue\n",
       "                    <numeric>      <numeric> <numeric> <numeric> <numeric>\n",
       "ENSG00000223972.4   0.2741718      0.2933056  5.201254 0.0563913  0.955030\n",
       "ENSG00000227232.4 226.0935890      0.0277149  0.242435 0.1143193  0.908985\n",
       "ENSG00000243485.2   0.0421757      0.2933932  5.201254 0.0564082  0.955017\n",
       "ENSG00000237613.2   0.0000000             NA        NA        NA        NA\n",
       "ENSG00000268020.2   0.0000000             NA        NA        NA        NA\n",
       "ENSG00000240361.1   0.0000000             NA        NA        NA        NA\n",
       "                       padj\n",
       "                  <numeric>\n",
       "ENSG00000223972.4        NA\n",
       "ENSG00000227232.4  0.968508\n",
       "ENSG00000243485.2        NA\n",
       "ENSG00000237613.2        NA\n",
       "ENSG00000268020.2        NA\n",
       "ENSG00000240361.1        NA"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#### Run DESeq2 and obtain results table\n",
    "\n",
    "dds <- DESeq(dds)\n",
    "res <- results(dds)\n",
    "head(res)\n",
    "\n",
    "#### Remove version number from Ensembl geneid\n",
    "tmp=gsub(\"\\\\..*\",\"\",row.names(res))\n",
    "rownames(res) <- tmp\n",
    "\n",
    "#Add gene symbol\n",
    "\n",
    "res$symbol <-mapIds(org.Hs.eg.db, keys=rownames(res), column=\"SYMBOL\", keytype=\"ENSEMBL\")\n",
    "head(res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3484418c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract normalised counts\n",
    "\n",
    "normalized_counts <- counts(dds, normalized=TRUE)\n",
    "\n",
    "tmp=gsub(\"\\\\..*\",\"\",row.names(normalized_counts))\n",
    "rownames(normalized_counts) <- tmp\n",
    "\n",
    "write.csv (normalized_counts, file=\"Normalised_counts_04-04-2023_whole_dataset.csv\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe22d51e",
   "metadata": {},
   "source": [
    "# Obtain results for each desired contrast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "f2aa672a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "out of 38815 with nonzero total read count\n",
      "adjusted p-value < 0.05\n",
      "LFC > 0 (up)       : 795, 2%\n",
      "LFC < 0 (down)     : 710, 1.8%\n",
      "outliers [1]       : 7, 0.018%\n",
      "low counts [2]     : 19795, 51%\n",
      "(mean count < 3)\n",
      "[1] see 'cooksCutoff' argument of ?results\n",
      "[2] see 'independentFiltering' argument of ?results\n",
      "\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in .testForValidKeys(x, keys, keytype, fks): None of the keys entered are valid keys for 'ENSEMBL'. Please use the keys method to see a listing of valid arguments.\n",
     "output_type": "error",
     "traceback": [
      "Error in .testForValidKeys(x, keys, keytype, fks): None of the keys entered are valid keys for 'ENSEMBL'. Please use the keys method to see a listing of valid arguments.\nTraceback:\n",
      "1. mapIds(org.Hs.eg.db, keys = rownames(res_fetal.control_vs_paed.control), \n .     column = \"SYMBOL\", keytype = \"ENSEMBL\")",
      "2. mapIds(org.Hs.eg.db, keys = rownames(res_fetal.control_vs_paed.control), \n .     column = \"SYMBOL\", keytype = \"ENSEMBL\")",
      "3. mapIds_base(x, keys, column, keytype, ..., multiVals = multiVals)",
      "4. select(x, keys = unique(keys), columns = column, keytype = keytype)",
      "5. select(x, keys = unique(keys), columns = column, keytype = keytype)",
      "6. .select(x, keys, columns, keytype, jointype = jointype, ...)",
      "7. testSelectArgs(x, keys = keys, cols = cols, keytype = keytype, \n .     fks = fks, skipValidKeysTest = skipValidKeysTest)",
      "8. .testForValidKeys(x, keys, keytype, fks)",
      "9. stop(msg)"
     ]
    }
   ],
   "source": [
    "#### Obtain results for each desired contrast ####\n",
    "\n",
    "\n",
    "#### Fetal vs paed baseline DGE ####\n",
    "# NB Paed control is baseline here, so LFC >0 means expression is upreg in fetal compared to paed \n",
    "\n",
    "res_fetal.control_vs_paed.control <- results(dds, contrast=c(\"group\",\"fetal.control\",\"paed.control\"), alpha=0.05)\n",
    "summary(res_fetal.control_vs_paed.control)\n",
    "\n",
    "\n",
    "##### FETAL ONLY #####\n",
    "\n",
    "### Fetal control vs 6h_treated\n",
    "\n",
    "res_fetal.control_vs_fetal.6h <- results(dds, contrast=c(\"group\",\"fetal.6h\",\"fetal.control\"), alpha=0.05)\n",
    "summary(res_fetal.control_vs_fetal.6h)\n",
    "\n",
    "### Fetal control vs 24h_treated\n",
    "\n",
    "res_fetal.control_vs_fetal.24h <- results(dds, contrast=c(\"group\",\"fetal.24h\",\"fetal.control\"), alpha=0.05)\n",
    "summary(res_fetal.control_vs_fetal.24h)\n",
    "\n",
    "### Fetal control vs 48h_treated\n",
    "\n",
    "res_fetal.control_vs_fetal.48h <- results(dds, contrast=c(\"group\",\"fetal.48h\",\"fetal.control\"), alpha=0.05)\n",
    "summary(res_fetal.control_vs_fetal.48h)\n",
    "\n",
    "### Fetal control vs 96h_treated\n",
    "\n",
    "res_fetal.control_vs_fetal.96h <- results(dds, contrast=c(\"group\",\"fetal.96h\",\"fetal.control\"), alpha=0.05)\n",
    "summary(res_fetal.control_vs_fetal.96h)\n",
    "\n",
    "\n",
    "##### PAED ONLY #####\n",
    "\n",
    "### Paed control vs 6h_treated\n",
    "\n",
    "res_paed.control_vs_paed.6h <- results(dds, contrast=c(\"group\",\"paed.6h\",\"paed.control\"), alpha=0.05)\n",
    "summary(res_paed.control_vs_paed.6h)\n",
    "\n",
    "### Paed control vs 24h_treated\n",
    "\n",
    "res_paed.control_vs_paed.24h <- results(dds, contrast=c(\"group\",\"paed.24h\",\"paed.control\"), alpha=0.05)\n",
    "summary(res_paed.control_vs_paed.24h)\n",
    "\n",
    "### Paed control vs 48h_treated \n",
    "\n",
    "res_paed.control_vs_paed.48h <- results(dds, contrast=c(\"group\",\"paed.48h\",\"paed.control\"), alpha=0.05)\n",
    "summary(res_paed.control_vs_paed.48h)\n",
    "\n",
    "### Paed control vs 96h_treated#\n",
    "\n",
    "res_paed.control_vs_paed.96h <- results(dds, contrast=c(\"group\",\"paed.96h\",\"paed.control\"), alpha=0.05)\n",
    "summary(res_paed.control_vs_paed.96h)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "1ed4fee3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#### Add gene symbol as a column\n",
    "\n",
    "tmp=gsub(\"\\\\..*\",\"\",row.names(res_fetal.control_vs_paed.control))\n",
    "rownames(res_fetal.control_vs_paed.control) <- tmp\n",
    "\n",
    "res_fetal.control_vs_paed.control$symbol <-mapIds(org.Hs.eg.db, keys=rownames(res_fetal.control_vs_paed.control), column=\"SYMBOL\", keytype=\"ENSEMBL\")\n",
    "\n",
    "\n",
    "tmp=gsub(\"\\\\..*\",\"\",row.names(res_fetal.control_vs_fetal.6h))\n",
    "rownames(res_fetal.control_vs_fetal.6h) <- tmp\n",
    "\n",
    "res_fetal.control_vs_fetal.6h$symbol <-mapIds(org.Hs.eg.db, keys=rownames(res_fetal.control_vs_fetal.6h), column=\"SYMBOL\", keytype=\"ENSEMBL\")\n",
    "\n",
    "\n",
    "tmp=gsub(\"\\\\..*\",\"\",row.names(res_fetal.control_vs_fetal.24h))\n",
    "rownames(res_fetal.control_vs_fetal.24h) <- tmp\n",
    "\n",
    "res_fetal.control_vs_fetal.24h$symbol <-mapIds(org.Hs.eg.db, keys=rownames(res_fetal.control_vs_fetal.24h), column=\"SYMBOL\", keytype=\"ENSEMBL\")\n",
    "\n",
    "\n",
    "tmp=gsub(\"\\\\..*\",\"\",row.names(res_fetal.control_vs_fetal.48h))\n",
    "rownames(res_fetal.control_vs_fetal.48h) <- tmp\n",
    "\n",
    "res_fetal.control_vs_fetal.48h$symbol <-mapIds(org.Hs.eg.db, keys=rownames(res_fetal.control_vs_fetal.48h), column=\"SYMBOL\", keytype=\"ENSEMBL\")\n",
    "\n",
    "\n",
    "tmp=gsub(\"\\\\..*\",\"\",row.names(res_fetal.control_vs_fetal.96h))\n",
    "rownames(res_fetal.control_vs_fetal.96h) <- tmp\n",
    "\n",
    "res_fetal.control_vs_fetal.96h$symbol <-mapIds(org.Hs.eg.db, keys=rownames(res_fetal.control_vs_fetal.96h), column=\"SYMBOL\", keytype=\"ENSEMBL\")\n",
    "\n",
    "\n",
    "tmp=gsub(\"\\\\..*\",\"\",row.names(res_paed.control_vs_paed.6h))\n",
    "rownames(res_paed.control_vs_paed.6h) <- tmp\n",
    "\n",
    "res_paed.control_vs_paed.6h$symbol <-mapIds(org.Hs.eg.db, keys=rownames(res_paed.control_vs_paed.6h), column=\"SYMBOL\", keytype=\"ENSEMBL\")\n",
    "\n",
    "\n",
    "\n",
    "tmp=gsub(\"\\\\..*\",\"\",row.names(res_paed.control_vs_paed.24h))\n",
    "rownames(res_paed.control_vs_paed.24h) <- tmp\n",
    "\n",
    "res_paed.control_vs_paed.24h$symbol <-mapIds(org.Hs.eg.db, keys=rownames(res_paed.control_vs_paed.24h), column=\"SYMBOL\", keytype=\"ENSEMBL\")\n",
    "\n",
    "\n",
    "\n",
    "tmp=gsub(\"\\\\..*\",\"\",row.names(res_paed.control_vs_paed.48h))\n",
    "rownames(res_paed.control_vs_paed.48h) <- tmp\n",
    "\n",
    "res_paed.control_vs_paed.48h$symbol <-mapIds(org.Hs.eg.db, keys=rownames(res_paed.control_vs_paed.48h), column=\"SYMBOL\", keytype=\"ENSEMBL\")\n",
    "\n",
    "\n",
    "tmp=gsub(\"\\\\..*\",\"\",row.names(res_paed.control_vs_paed.96h))\n",
    "rownames(res_paed.control_vs_paed.96h) <- tmp\n",
    "\n",
    "res_paed.control_vs_paed.96h$symbol <-mapIds(org.Hs.eg.db, keys=rownames(res_paed.control_vs_paed.96h), column=\"SYMBOL\", keytype=\"ENSEMBL\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "236b4d88",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "log2 fold change (MLE): group fetal.control vs paed.control \n",
       "Wald test p-value: group fetal.control vs paed.control \n",
       "DataFrame with 6 rows and 7 columns\n",
       "                 baseMean log2FoldChange     lfcSE      stat      pvalue\n",
       "                <numeric>      <numeric> <numeric> <numeric>   <numeric>\n",
       "ENSG00000162896 20418.514       -8.24085  0.404743  -20.3607 3.73286e-92\n",
       "ENSG00000130600  1962.064       10.27065  0.559672   18.3512 3.22931e-75\n",
       "ENSG00000109113  1084.542        5.59643  0.387288   14.4503 2.49634e-47\n",
       "ENSG00000165973   412.849        7.49469  0.532068   14.0860 4.63233e-45\n",
       "ENSG00000159217   473.191        8.18067  0.616383   13.2721 3.36181e-40\n",
       "ENSG00000185615   120.466       -5.84376  0.448589  -13.0270 8.59427e-39\n",
       "                       padj      symbol\n",
       "                  <numeric> <character>\n",
       "ENSG00000162896 7.09728e-88        PIGR\n",
       "ENSG00000130600 3.06994e-71         H19\n",
       "ENSG00000109113 1.58210e-43       RAB34\n",
       "ENSG00000165973 2.20186e-41       NELL1\n",
       "ENSG00000159217 1.27836e-36     IGF2BP1\n",
       "ENSG00000185615 2.72338e-35       PDIA2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "##### Fetal control vs paed control ####\n",
    "#### Obtain just those genes with p-adj <0.05 ###\n",
    "\n",
    "### Reorder and get the list of Differentially Expressed Genes (DEGs) p-values and adjusted p-values\n",
    "\n",
    "resOrdered_fetal.control_vs_paed.control <- res_fetal.control_vs_paed.control[order(res_fetal.control_vs_paed.control$padj),]\n",
    "\n",
    "### Retain DEG with padj < 0.05 and exporting as .csv file\n",
    "resSig_fetal.control_vs_paed.control <- subset(resOrdered_fetal.control_vs_paed.control, padj < 0.05)\n",
    "head(resSig_fetal.control_vs_paed.control)\n",
    "\n",
    "write.csv(as.data.frame(resSig_fetal.control_vs_paed.control), file=\"DEGs_fetal_control_vs_paed_control.csv\")\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "26769b97",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "log2 fold change (MLE): group fetal.6h vs fetal.control \n",
       "Wald test p-value: group fetal.6h vs fetal.control \n",
       "DataFrame with 6 rows and 7 columns\n",
       "                 baseMean log2FoldChange     lfcSE      stat      pvalue\n",
       "                <numeric>      <numeric> <numeric> <numeric>   <numeric>\n",
       "ENSG00000161547  1649.868        1.50572  0.198229   7.59586 3.05746e-14\n",
       "ENSG00000121390   820.951        1.27418  0.183217   6.95448 3.53872e-12\n",
       "ENSG00000136527  3391.467        1.15903  0.179253   6.46590 1.00698e-10\n",
       "ENSG00000110172   907.121        1.75111  0.271889   6.44054 1.19046e-10\n",
       "ENSG00000173145   619.859        1.45133  0.251855   5.76255 8.28542e-09\n",
       "ENSG00000185361   554.287       -1.27112  0.221000  -5.75166 8.83712e-09\n",
       "                       padj      symbol\n",
       "                  <numeric> <character>\n",
       "ENSG00000161547 4.02056e-10       SRSF2\n",
       "ENSG00000121390 2.32671e-08       PSPC1\n",
       "ENSG00000136527 3.91362e-07       TRA2B\n",
       "ENSG00000110172 3.91362e-07     CHORDC1\n",
       "ENSG00000173145 1.93680e-05       NOC3L\n",
       "ENSG00000185361 1.93680e-05   TNFAIP8L1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "###### FETAL ONLY DEG lists ######\n",
    "##### Fetal control vs 6hr ######\n",
    "#### Obtain just those genes with p-adj <0.05 ###\n",
    "\n",
    "### Reorder and get the list of Differentially Expressed Genes (DEGs) p-values and adjusted p-values\n",
    "\n",
    "resOrdered_fetal.control_vs_fetal.6h <- res_fetal.control_vs_fetal.6h[order(res_fetal.control_vs_fetal.6h$padj),]\n",
    "\n",
    "### Retain DEG with padj < 0.05 and exporting as .csv file\n",
    "resSig_fetal.control_vs_fetal.6h <- subset(resOrdered_fetal.control_vs_fetal.6h, padj < 0.05)\n",
    "head(resSig_fetal.control_vs_fetal.6h)\n",
    "\n",
    "write.csv(as.data.frame(resSig_fetal.control_vs_fetal.6h), file=\"DEGs_fetal_control_vs_fetal_6h.csv\")\n",
    "\n",
    "\n",
    "##### Fetal control vs 24hr #####\n",
    "#### Obtain just those genes with p-adj <0.05 ###\n",
    "\n",
    "### Reorder and get the list of Differentially Expressed Genes (DEGs) p-values and adjusted p-values\n",
    "\n",
    "resOrdered_fetal.control_vs_fetal.24h <- res_fetal.control_vs_fetal.24h[order(res_fetal.control_vs_fetal.24h$padj),]\n",
    "\n",
    "### Retain DEG with padj < 0.05 and exporting as .csv file\n",
    "resSig_fetal.control_vs_fetal.24h <- subset(resOrdered_fetal.control_vs_fetal.24h, padj < 0.05)\n",
    "head(resSig_fetal.control_vs_fetal.24h)\n",
    "\n",
    "write.csv(as.data.frame(resSig_fetal.control_vs_fetal.24h), file=\"DEGs_fetal_control_vs_fetal_24h.csv\")\n",
    "\n",
    "\n",
    "##### Fetal control vs 48hr #####\n",
    "#### Obtain just those genes with p-adj <0.05 ###\n",
    "\n",
    "### Reorder and get the list of Differentially Expressed Genes (DEGs) p-values and adjusted p-values\n",
    "\n",
    "resOrdered_fetal.control_vs_fetal.48h <- res_fetal.control_vs_fetal.48h[order(res_fetal.control_vs_fetal.48h$padj),]\n",
    "\n",
    "### Retain DEG with padj < 0.05 and exporting as .csv file\n",
    "resSig_fetal.control_vs_fetal.48h <- subset(resOrdered_fetal.control_vs_fetal.48h, padj < 0.05)\n",
    "head(resSig_fetal.control_vs_fetal.48h)\n",
    "\n",
    "write.csv(as.data.frame(resSig_fetal.control_vs_fetal.48h), file=\"DEGs_fetal_control_vs_fetal_48h.csv\")\n",
    "\n",
    "\n",
    "##### Fetal control vs 96hr #####\n",
    "#### Obtain just those genes with p-adj <0.05 ###\n",
    "\n",
    "### Reorder and get the list of Differentially Expressed Genes (DEGs) p-values and adjusted p-values\n",
    "\n",
    "resOrdered_fetal.control_vs_fetal.96h <- res_fetal.control_vs_fetal.96h[order(res_fetal.control_vs_fetal.96h$padj),]\n",
    "\n",
    "### Retain DEG with padj < 0.05 and exporting as .csv file\n",
    "resSig_fetal.control_vs_fetal.96h <- subset(resOrdered_fetal.control_vs_fetal.96h, padj < 0.05)\n",
    "head(resSig_fetal.control_vs_fetal.96h)\n",
    "\n",
    "write.csv(as.data.frame(resSig_fetal.control_vs_fetal.96h), file=\"DEGs_fetal_control_vs_fetal_96h.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "ad3a3fd6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "log2 fold change (MLE): group paed.6h vs paed.control \n",
       "Wald test p-value: group paed.6h vs paed.control \n",
       "DataFrame with 6 rows and 7 columns\n",
       "                 baseMean log2FoldChange     lfcSE      stat      pvalue\n",
       "                <numeric>      <numeric> <numeric> <numeric>   <numeric>\n",
       "ENSG00000185361   554.287      -1.487625  0.223800  -6.64713 2.98868e-11\n",
       "ENSG00000126351   619.651      -0.967631  0.170718  -5.66801 1.44470e-08\n",
       "ENSG00000163486   233.610      -1.806457  0.342094  -5.28059 1.28767e-07\n",
       "ENSG00000072201   157.511      -1.714654  0.328165  -5.22497 1.74180e-07\n",
       "ENSG00000204388   656.853       1.403169  0.266297   5.26919 1.37030e-07\n",
       "ENSG00000165055   620.258       1.086329  0.206798   5.25309 1.49567e-07\n",
       "                       padj      symbol\n",
       "                  <numeric> <character>\n",
       "ENSG00000185361 3.93011e-07   TNFAIP8L1\n",
       "ENSG00000126351 9.49889e-05        THRA\n",
       "ENSG00000163486 3.27209e-04          NA\n",
       "ENSG00000072201 3.27209e-04        LNX1\n",
       "ENSG00000204388 3.27209e-04      HSPA1B\n",
       "ENSG00000165055 3.27209e-04     METTL2B"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "####### PAED ONLY DEG LISTS ########\n",
    "\n",
    "##### Paed control vs 6hr ######\n",
    "#### Obtain just those genes with p-adj <0.05 ###\n",
    "\n",
    "### Reorder and get the list of Differentially Expressed Genes (DEGs) p-values and adjusted p-values\n",
    "\n",
    "resOrdered_paed.control_vs_paed.6h <- res_paed.control_vs_paed.6h[order(res_paed.control_vs_paed.6h$padj),]\n",
    "\n",
    "### Retain DEG with padj < 0.05 and exporting as .csv file\n",
    "resSig_paed.control_vs_paed.6h <- subset(resOrdered_paed.control_vs_paed.6h, padj < 0.05)\n",
    "head(resSig_paed.control_vs_paed.6h)\n",
    "\n",
    "write.csv(as.data.frame(resSig_paed.control_vs_paed.6h), file=\"DEGs_paed_control_vs_paed_6h.csv\")\n",
    "\n",
    "\n",
    "##### Paed control vs 24hr #####\n",
    "#### Obtain just those genes with p-adj <0.05 ###\n",
    "\n",
    "### Reorder and get the list of Differentially Expressed Genes (DEGs) p-values and adjusted p-values\n",
    "\n",
    "resOrdered_paed.control_vs_paed.24h <- res_paed.control_vs_paed.24h[order(res_paed.control_vs_paed.24h$padj),]\n",
    "\n",
    "### Retain DEG with padj < 0.05 and exporting as .csv file\n",
    "resSig_paed.control_vs_paed.24h <- subset(resOrdered_paed.control_vs_paed.24h, padj < 0.05)\n",
    "head(resSig_paed.control_vs_paed.24h)\n",
    "\n",
    "write.csv(as.data.frame(resSig_paed.control_vs_paed.24h), file=\"DEGs_paed_control_vs_paed_24h.csv\")\n",
    "\n",
    "\n",
    "###### Paed control vs 48h #######\n",
    "#### Obtain just those genes with p-adj <0.05 ###\n",
    "\n",
    "### Reorder and get the list of Differentially Expressed Genes (DEGs) p-values and adjusted p-values\n",
    "\n",
    "resOrdered_paed.control_vs_paed.48h <- res_paed.control_vs_paed.48h[order(res_paed.control_vs_paed.48h$padj),]\n",
    "\n",
    "### Retain DEG with padj < 0.05 and exporting as .csv file\n",
    "resSig_paed.control_vs_paed.48h <- subset(resOrdered_paed.control_vs_paed.48h, padj < 0.05)\n",
    "head(resSig_paed.control_vs_paed.48h)\n",
    "\n",
    "write.csv(as.data.frame(resSig_paed.control_vs_paed.48h), file=\"DEGs_paed_control_vs_paed_48h.csv\")\n",
    "\n",
    "\n",
    "###### Paed control vs 96hr ######\n",
    "#### Obtain just those genes with p-adj <0.05 ###\n",
    "\n",
    "### Reorder and get the list of Differentially Expressed Genes (DEGs) p-values and adjusted p-values\n",
    "\n",
    "resOrdered_paed.control_vs_paed.96h <- res_paed.control_vs_paed.96h[order(res_paed.control_vs_paed.96h$padj),]\n",
    "\n",
    "### Retain DEG with padj < 0.05 and exporting as .csv file\n",
    "resSig_paed.control_vs_paed.96h <- subset(resOrdered_paed.control_vs_paed.96h, padj < 0.05)\n",
    "head(resSig_paed.control_vs_paed.96h)\n",
    "\n",
    "write.csv(as.data.frame(resSig_paed.control_vs_paed.96h), file=\"DEGs_paed_control_vs_paed_96h.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "236807de",
   "metadata": {},
   "source": [
    "# LOOKING FOR JOINT AND UNIQUE HITS BETWEEN FETAL AND PAEDIATRIC SIG DEGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "3f3725e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "125"
      ],
      "text/latex": [
       "125"
      ],
      "text/markdown": [
       "125"
      ],
      "text/plain": [
       "[1] 125"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "220"
      ],
      "text/latex": [
       "220"
      ],
      "text/markdown": [
       "220"
      ],
      "text/plain": [
       "[1] 220"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "1"
      ],
      "text/latex": [
       "1"
      ],
      "text/markdown": [
       "1"
      ],
      "text/plain": [
       "[1] 1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "2"
      ],
      "text/latex": [
       "2"
      ],
      "text/markdown": [
       "2"
      ],
      "text/plain": [
       "[1] 2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "363"
      ],
      "text/latex": [
       "363"
      ],
      "text/markdown": [
       "363"
      ],
      "text/plain": [
       "[1] 363"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "401"
      ],
      "text/latex": [
       "401"
      ],
      "text/markdown": [
       "401"
      ],
      "text/plain": [
       "[1] 401"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "987"
      ],
      "text/latex": [
       "987"
      ],
      "text/markdown": [
       "987"
      ],
      "text/plain": [
       "[1] 987"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "1081"
      ],
      "text/latex": [
       "1081"
      ],
      "text/markdown": [
       "1081"
      ],
      "text/plain": [
       "[1] 1081"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## Joint and Unqiue hits ##\n",
    "\n",
    "### 6hr timepoint ###\n",
    "\n",
    "# Make a shorter name\n",
    "resSig_fetal.control_vs_fetal.6h -> DEGs_fetal_6h\n",
    "resSig_paed.control_vs_paed.6h -> DEGs_paed_6h\n",
    "\n",
    "# Make geneid a 'proper' column in DEGs table\n",
    "DEGs_fetal_6h$geneid <- rownames(DEGs_fetal_6h)\n",
    "DEGs_paed_6h$geneid <- rownames(DEGs_paed_6h)\n",
    "\n",
    "# Looking for number of joint hits between fetal and paed samples out of DEGs identified in control vs 6hr DGE analyses\n",
    "\n",
    "overlap <- subset(DEGs_fetal_6h, geneid %in% DEGs_paed_6h$geneid)\n",
    "nrow(overlap)\n",
    "\n",
    "# Looking for number of unique hits in fetal samples (i.e. gene hits that are not in the paed samples\n",
    "\n",
    "missing.1 <- subset(DEGs_fetal_6h, !(geneid %in% DEGs_paed_6h$geneid))\n",
    "nrow(missing.1)\n",
    "\n",
    "# Looking for number of unique hits in paed samples (i.e. gene hits that are not in fetal samples)\n",
    "\n",
    "missing.2 <- subset(DEGs_paed_6h, !(geneid %in% DEGs_fetal_6h$geneid))\n",
    "\n",
    "### Export the hits unique to fetal samples ###\n",
    "\n",
    "write.csv(as.data.frame(missing.1), file=\"DEGs_6h_unique_to_fetal.csv\")\n",
    "\n",
    "## Export joint hits at 6hr\n",
    "\n",
    "write.csv(as.data.frame(overlap), file=\"DEGs_6h_joint_hits.csv\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "### 24hr timepoint ###\n",
    "\n",
    "# Make a shorter name\n",
    "resSig_fetal.control_vs_fetal.24h -> DEGs_fetal_24h\n",
    "resSig_paed.control_vs_paed.24h -> DEGs_paed_24h\n",
    "\n",
    "# Make geneid a 'proper' column in DEGs table\n",
    "DEGs_fetal_24h$geneid <- rownames(DEGs_fetal_24h)\n",
    "DEGs_paed_24h$geneid <- rownames(DEGs_paed_24h)\n",
    "\n",
    "# Looking for number of joint hits between fetal and paed samples out of DEGs identified in control vs 6hr DGE analyses\n",
    "\n",
    "overlap <- subset(DEGs_fetal_24h, geneid %in% DEGs_paed_24h$geneid)\n",
    "nrow(overlap)\n",
    "\n",
    "# Looking for number of unique hits in fetal samples (i.e. gene hits that are not in the paed samples\n",
    "\n",
    "missing.1 <- subset(DEGs_fetal_24h, !(geneid %in% DEGs_paed_24h$geneid))\n",
    "nrow(missing.1)\n",
    "\n",
    "# Looking for number of unique hits in paed samples (i.e. gene hits that are not in fetal samples)\n",
    "\n",
    "missing.2 <- subset(DEGs_paed_24h, !(geneid %in% DEGs_fetal_24h$geneid))\n",
    "\n",
    "### Export the hits unique to fetal samples ###\n",
    "\n",
    "write.csv(as.data.frame(missing.1), file=\"DEGs_24h_unique_to_fetal.csv\")\n",
    "\n",
    "## Export joint hits at 24hr\n",
    "\n",
    "write.csv(as.data.frame(overlap), file=\"DEGs_24h_joint_hits.csv\")\n",
    "\n",
    "## Export hits unique to paed samples \n",
    "\n",
    "write.csv(as.data.frame(missing.2), file= \"DEGs_24h_unique_to_paed.csv\")\n",
    "\n",
    "\n",
    "\n",
    "#### 48hr timepoint ####\n",
    "\n",
    "# Make a shorter name\n",
    "resSig_fetal.control_vs_fetal.48h -> DEGs_fetal_48h\n",
    "resSig_paed.control_vs_paed.48h -> DEGs_paed_48h\n",
    "\n",
    "# Make geneid a 'proper' column in DEGs table\n",
    "DEGs_fetal_48h$geneid <- rownames(DEGs_fetal_48h)\n",
    "DEGs_paed_48h$geneid <- rownames(DEGs_paed_48h)\n",
    "\n",
    "# Looking for number of joint hits between fetal and paed samples out of DEGs identified in control vs 6hr DGE analyses\n",
    "\n",
    "overlap <- subset(DEGs_fetal_48h, geneid %in% DEGs_paed_48h$geneid)\n",
    "nrow(overlap)\n",
    "\n",
    "# Looking for number of unique hits in fetal samples (i.e. gene hits that are not in the paed samples\n",
    "\n",
    "missing.1 <- subset(DEGs_fetal_48h, !(geneid %in% DEGs_paed_48h$geneid))\n",
    "nrow(missing.1)\n",
    "\n",
    "# Looking for number of unique hits in paed samples (i.e. gene hits that are not in fetal samples)\n",
    "\n",
    "missing.2 <- subset(DEGs_paed_48h, !(geneid %in% DEGs_fetal_48h$geneid))\n",
    "\n",
    "### Export the hits unique to fetal samples ###\n",
    "\n",
    "write.csv(as.data.frame(missing.1), file=\"DEGs_48h_unique_to_fetal.csv\")\n",
    "\n",
    "## Export joint hits at 48hr\n",
    "\n",
    "write.csv(as.data.frame(overlap), file=\"DEGs_48h_joint_hits.csv\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#### 96hr timepoint ####\n",
    "\n",
    "# Make a shorter name\n",
    "resSig_fetal.control_vs_fetal.96h -> DEGs_fetal_96h\n",
    "resSig_paed.control_vs_paed.96h -> DEGs_paed_96h\n",
    "\n",
    "# Make geneid a 'proper' column in DEGs table\n",
    "DEGs_fetal_96h$geneid <- rownames(DEGs_fetal_96h)\n",
    "DEGs_paed_96h$geneid <- rownames(DEGs_paed_96h)\n",
    "\n",
    "# Looking for number of joint hits between fetal and paed samples out of DEGs identified in control vs 96hr DGE analyses\n",
    "\n",
    "overlap <- subset(DEGs_fetal_96h, geneid %in% DEGs_paed_96h$geneid)\n",
    "nrow(overlap)\n",
    "\n",
    "# Looking for number of unique hits in fetal samples (i.e. gene hits that are not in the paed samples\n",
    "\n",
    "missing.1 <- subset(DEGs_fetal_96h, !(geneid %in% DEGs_paed_96h$geneid))\n",
    "nrow(missing.1)\n",
    "\n",
    "# Looking for number of unique hits in paed samples (i.e. gene hits that are not in fetal samples)\n",
    "\n",
    "missing.2 <- subset(DEGs_paed_96h, !(geneid %in% DEGs_fetal_96h$geneid))\n",
    "\n",
    "### Export the hits unique to fetal samples ###\n",
    "\n",
    "write.csv(as.data.frame(missing.1), file=\"DEGs_96h_unique_to_fetal.csv\")\n",
    "\n",
    "## Export joint hits at 48hr\n",
    "\n",
    "write.csv(as.data.frame(overlap), file=\"DEGs_96h_joint_hits.csv\")\n",
    "\n",
    "## Export hits unique to paed samples at 96hr timepoint \n",
    "write.csv(as.data.frame(missing.2), file=\"DEGs_96h_unique_to_paed.csv\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41129cc5",
   "metadata": {},
   "source": [
    "# Simple ggplot RNA-Seq normalised counts across timeseries \n",
    "### Control and treated, Fetal and Paediatric "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b961622",
   "metadata": {},
   "outputs": [],
   "source": [
    "######### MY ATTEMPT AT SIMPLE RNA-SEQ DATA PLOTS OVER TIME ########\n",
    "##### Load packages####\n",
    "library(ggpubr)\n",
    "install.packages(\"tidyverse\")\n",
    "library(tidyverse)\n",
    "library(dplyr)\n",
    "library(ggplot2)\n",
    "library(plyr)\n",
    "## Setting the working directory (add the path to where you store your normalised counts and simples information)\n",
    "setwd(\"C:/Users/issyl/OneDrive - University of Cambridge/Research Project/DGE analysis AAI/DESeq2 analysis/RNA-Seq-AAI-analysis/00_raw data\")\n",
    "getwd()\n",
    "\n",
    "## Importing the normalised counts and samples information\n",
    "#Let's assume you have a .csv file format\n",
    "\n",
    "counts <- read.csv(file.choose())\n",
    "head(counts)\n",
    "meta <- read.table(file.choose(), header=T, sep=\"\\t\", stringsAsFactors=F)\n",
    "head(meta)\n",
    "\n",
    "# add the rownames as a proper column\n",
    "#Let's set the ensembl ID as colnames for us to easily select/change features later\n",
    "\n",
    "colnames(counts)[1] <- \"feature\"\n",
    "head(counts)\n",
    "\n",
    "# Select your feature of interest/gene\n",
    "### p21 ###\n",
    "selected_p21 <- c(\"ENSG00000124762\")\n",
    "\n",
    "counts_selected <- counts[counts$feature %in% selected_p21, ] \n",
    "counts_selected \n",
    "\n",
    "# Generate a long format then combine the counts table and the sample information \n",
    "counts_long <- counts_selected  %>% gather(sample, count, 2:ncol(counts_selected ), factor_key = TRUE)\n",
    "counts_long\n",
    "\n",
    "\n",
    "# Merge the two tables together\n",
    "counts_long_annot <- merge(counts_long, meta, by = \"sample\")\n",
    "counts_long_annot\n",
    "\n",
    "\n",
    "# Arrange data based on a factor you want (e.g stage: fetal vs adult) then enter the levels in the order you want (e.g fetal then adult) \n",
    "#### STAGE ####\n",
    "counts_long_annot$stage <- factor(counts_long_annot$stage, levels = c(\"fetal\", \"adult\"))\n",
    "counts_long_annot\n",
    "\n",
    "#### GROUP ####\n",
    "counts_long_annot$group <- factor(counts_long_annot$group, levels = c(\"fetal.control\", \"fetal.6h\", \"fetal.24h\", \"fetal.48h\", \"fetal.96h\", \"paed.control\", \"paed.6h\", \"paed.24h\", \"paed.48h\", \"paed.96h\"))\n",
    "counts_long_annot\n",
    "\n",
    "#### LINE####\n",
    "counts_long_annot$line <- factor(counts_long_annot$line, levels = c(\"X93\", \"X96\", \"X647\", \"X655\"))\n",
    "counts_long_annot\n",
    "\n",
    "#### TIMEPOINT ####\n",
    "counts_long_annot$timepoint <- factor(counts_long_annot$timepoint, levels = c(\"0C\", \"48C\", \"6\", \"24\", \"48\", \"96\"))\n",
    "counts_long_annot\n",
    "\n",
    "\n",
    "# Changing name to a short one\n",
    "data <- counts_long_annot\n",
    "data\n",
    "\n",
    "\n",
    "## Use ggplot to plot the time series data I assume you will use 0 hours, 6hours, 24 hours, 48hours + 48h untreated then recovery\n",
    "\n",
    "ggplot(data = data, aes(x=timepoint, y=count, group=line, colour=stage)) + geom_line() + labs(y= \"Norm Counts\", x= \"Timepoint\") + ggtitle(\"p21 Expression levels timeseries\")\n",
    "\n",
    "\n",
    "### ggplot with 'shape' indicating the biological replicate and colour indicating stage ##\n",
    "gg <- ggplot(data = data) \n",
    "gg <- gg + geom_line(aes(x=timepoint, y=count, group=line, colour=stage))\n",
    "gg <- gg + geom_point(aes(x=timepoint, y=count, shape=line), size=1.5)\n",
    "gg <- gg + labs(y= \"Norm Counts\", x= \"Timepoint\")\n",
    "gg <- gg + ggtitle(\"WNT7A: Timecourse of Normalised Counts\")\n",
    "gg\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e28d5a66",
   "metadata": {},
   "source": [
    "# Heatmap script\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35263e4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "##### HEATMAP SCRIPT ####\n",
    "\n",
    "#### Load packages ####\n",
    "#install.packages(\"pheatmap\")\n",
    "require(pheatmap)\n",
    "#BiocManager::install(\"RColorBrewer\")\n",
    "library(RColorBrewer)\n",
    "library(tidyverse)\n",
    "library(dplyr)\n",
    "library(pheatmap)\n",
    "\n",
    "install.packages(\"colorRampPalette\")\n",
    "install.packages(\"Rtools\")\n",
    "library(AnnotationDbi)\n",
    "library(org.Hs.eg.db)\n",
    "\n",
    "### Set working directory ###\n",
    "\n",
    "setwd('C:/Users/issyl/OneDrive - University of Cambridge/Research Project/DGE analysis AAI/DESeq2 analysis/RNA-Seq-AAI-analysis/normalised counts/Norm counts whole dataset included')\n",
    "\n",
    "### Read in normalised counts data ###\n",
    "\n",
    "read.csv(\"USE_THIS_Normalised_counts_whole_dataset_low_counts_removed_5-4-23.csv\") -> norm_counts \n",
    "\n",
    "## Make geneid the rownames ##\n",
    "norm_counts$geneid -> rownames(norm_counts)\n",
    "head(norm_counts)\n",
    "\n",
    "### To make just one geneid column ##\n",
    "norm_counts <- norm_counts[, -1]\n",
    "head(norm_counts)\n",
    "\n",
    "### Add gene symbols ###\n",
    "norm_counts$symbol <-mapIds(org.Hs.eg.db, keys=rownames(norm_counts), column=\"SYMBOL\", keytype=\"ENSEMBL\")\n",
    "\n",
    "### Select GOIs to use in heatmap from normalised counts ###\n",
    "norm_counts[c('ENSG00000141510', 'ENSG00000135679', 'ENSG00000116717', 'ENSG00000134574', 'ENSG00000087088', 'ENSG00000124762', 'ENSG00000110092' ),] -> p53_GOIs\n",
    "\n",
    "## Make the gene symbols the row names ##\n",
    "p53_GOIs$symbol -> rownames(p53_GOIs)\n",
    "\n",
    "## Remove other gene symbol column ##\n",
    "p53_GOIs <- p53_GOIs[, -25]\n",
    "head(p53_GOIs)\n",
    "\n",
    "## Reorder samples to desired order for heatmap ##\n",
    "p53_GOIs <- p53_GOIs[, c(1, 2, 7, 8, 3, 9, 4, 10, 5, 11, 6, 12, 13, 14, 19, 20, 15, 21, 16, 22, 17, 23, 18, 24)]\n",
    "head(p53_GOIs)\n",
    "\n",
    "### Make it a matrix ###\n",
    "as.matrix(p53_GOIs) -> p53_GOIs\n",
    "\n",
    "#### Need an annotation file too ###\n",
    "setwd(\"C:/Users/issyl/OneDrive - University of Cambridge/Research Project/DGE analysis AAI/DESeq2 analysis/RNA-Seq-AAI-analysis/00_raw data\")\n",
    "read.delim(\"metadataRNAseqAAI.txt\") -> anno\n",
    "rownames(anno) <- anno$sample\n",
    "anno$sample <- NULL\n",
    "anno[, c(1, 2, 3)] -> anno\n",
    "\n",
    "### Make annotation file order match the counts data order ###\n",
    "anno <- anno[c(1, 2, 7, 8, 3, 9, 4, 10, 5, 11, 6, 12, 13, 14, 19, 20, 15, 21, 16, 22, 17, 23, 18, 24),]\n",
    "\n",
    "\n",
    "#### HEATMAP ###\n",
    "cols <- list(treatment=c(untreated=\"grey80\", treated= \"cornflowerblue\"), stage=c(fetal=\"grey60\", adult=\"aquamarine4\"), timepoint=c())\n",
    "\n",
    "heat <- pheatmap(p53_GOIs, color=colorRampPalette(c(\"blue\",\"white\",\"red3\"))(50), clustering_distance_cols=\"euclidean\", clustering_method=\"complete\", cluster_rows=F, cluster_cols=F, treeheight_row=10, treeheight_col=50, cellheight=15, cellwidth=15, border_color=F, scale=\"row\", show_colnames=T, show_rownames=T, annotation_col=anno, annotation_colors=cols)\n",
    "\n",
    "\n",
    "##### Heatmap script using a gene list to select genes of interest regarding a particular pathway/GO term ####\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bc73012",
   "metadata": {},
   "source": [
    "# DESeq2 analysis for volcano plots and GO enrichment plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a089368",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### DESeq2 analysis: combining 24hr, 48hr and 96hr into one 'group' \n",
    "and doing DGE analysis between untreated (0C and 48C) and treated (24+48) ###\n",
    "\n",
    "#### Set working directory ####\n",
    "setwd('C:/Users/issyl/OneDrive - University of Cambridge/Research Project/DGE analysis AAI/DESeq2 analysis/RNA-Seq-AAI-analysis/00_raw data')\n",
    "\n",
    "## Loading counts matrix and metadata\n",
    "\n",
    "cnts <- read.delim(file.choose(), row.names=1)\n",
    "coldata<- read.delim(file.choose(), row.names=1)\n",
    "\n",
    "\n",
    "### Use this for patients study samples\n",
    "coldata <- coldata[,c(\"treatment\", \"timepoint\", \"stage\", \"line\", \"group\")]\n",
    "\n",
    "# Make factors \n",
    "coldata$treatment <- factor(coldata$treatment)\n",
    "coldata$timepoint <- factor(coldata$timepoint)\n",
    "coldata$stage <- factor(coldata$stage)\n",
    "coldata$ID <- factor(coldata$ID)\n",
    "coldata$line <- factor(coldata$line)\n",
    "coldata$group <- factor(coldata$group)\n",
    "\n",
    "\n",
    "### Make a DESeq2 design \n",
    "dds <- DESeqDataSetFromMatrix(countData = cnts,\n",
    "                              colData = coldata,\n",
    "                              design = ~ group)\n",
    "\n",
    "#### Set factor level, have set fetal.control as reference###\n",
    "dds$group <- relevel(dds$group, ref=\"0\")\n",
    "\n",
    "### Filter out feature with very low counts and keep only feature with counts' sum of =>10\n",
    "keep <- rowSums(counts(dds)) >= 10\n",
    "dds <- dds[keep,]\n",
    "\n",
    "#### Run DESeq2 and obtain results table\n",
    "\n",
    "dds <- DESeq(dds)\n",
    "res <- results(dds)\n",
    "head(res)\n",
    "\n",
    "#### Remove version number from Ensembl geneid\n",
    "tmp=gsub(\"\\\\..*\",\"\",row.names(res))\n",
    "rownames(res) <- tmp\n",
    "\n",
    "#Add gene symbol\n",
    "\n",
    "res$symbol <-mapIds(org.Hs.eg.db, keys=rownames(res), column=\"SYMBOL\", keytype=\"ENSEMBL\")\n",
    "head(res)\n",
    "\n",
    "write.csv(res, file= \"full_results_foetal_group_0_24h48h.csv\")\n",
    "\n",
    "#### Use contrast() function to do paed.ctrl vs 24+48h treated and foetal.ctrl vs 24+48h treated ####\n",
    "## Then extract list of DEGs for enrichment analysis SHINY GO ####\n",
    "\n",
    "#### PAED\n",
    "\n",
    "res_paed_ctrl_vs_24h48h <- results(dds, contrast=c(\"group\",\"24h48h\",\"0\"), alpha=0.05)\n",
    "summary(res_paed_ctrl_vs_24h48h)\n",
    "head(res_paed_ctrl_vs_24h48h)\n",
    "\n",
    "\n",
    "tmp=gsub(\"\\\\..*\",\"\",row.names(res_paed_ctrl_vs_24h48h))\n",
    "rownames(res_paed_ctrl_vs_24h48h) <- tmp\n",
    "\n",
    "res_paed_ctrl_vs_24h48h$symbol <-mapIds(org.Hs.eg.db, keys=rownames(res_paed_ctrl_vs_24h48h), column=\"SYMBOL\", keytype=\"ENSEMBL\")\n",
    "\n",
    "#### Obtain just those genes with p-adj <0.05 ###\n",
    "\n",
    "### Reorder and get the list of Differentially Expressed Genes (DEGs) p-values and adjusted p-values\n",
    "\n",
    "resOrdered_paed_ctrl_vs_24h48h <- res_paed_ctrl_vs_24h48h[order(res_paed_ctrl_vs_24h48h$padj),]\n",
    "\n",
    "### Retain DEG with padj < 0.05 and exporting as .csv file\n",
    "resSig_paed_ctrl_vs_24h48h <- subset(resOrdered_paed_ctrl_vs_24h48h, padj < 0.05)\n",
    "head(resSig_paed_ctrl_vs_24h48h)\n",
    "\n",
    "write.csv(as.data.frame(resSig_paed_ctrl_vs_24h48h), file=\"DEGs_paed_ctrl_vs_24h48h.csv\")\n",
    "\n",
    "####Upreg sig DEGs ####\n",
    "sig_DEGs_paed_upreg_24h48h <- subset(resSig_paed_ctrl_vs_24h48h, log2FoldChange >= 0.5)\n",
    "summary(sig_DEGs_paed_upreg_24h48h)\n",
    "\n",
    "write.csv(as.data.frame(sig_DEGs_paed_upreg_24h48h), file=\"sig_upreg_DEGs_paed_24h48h.csv\")\n",
    "\n",
    "#### Downreg sig DEGs ####\n",
    "sig_DEGs_paed_downreg_24h48h <- subset(resSig_paed_ctrl_vs_24h48h, log2FoldChange <= -0.5)\n",
    "summary(sig_DEGs_paed_downreg_24h48h)\n",
    "\n",
    "write.csv(as.data.frame(sig_DEGs_paed_downreg_24h48h), file=\"sig_downreg_DEGs_paed_24h48h.csv\")\n",
    "\n",
    "### Foetal \n",
    "\n",
    "res_foetal_ctrl_vs_24h48h <- results(dds, contrast=c(\"group\",\"24h48h\",\"0\"), alpha=0.05)\n",
    "summary(res_foetal_ctrl_vs_24h48h)\n",
    "tmp=gsub(\"\\\\..*\",\"\",row.names(res_foetal_ctrl_vs_24h48h))\n",
    "rownames(res_foetal_ctrl_vs_24h48h) <- tmp\n",
    "\n",
    "res_foetal_ctrl_vs_24h48h$symbol <-mapIds(org.Hs.eg.db, keys=rownames(res_foetal_ctrl_vs_24h48h), column=\"SYMBOL\", keytype=\"ENSEMBL\")\n",
    "\n",
    "#### Obtain just those genes with p-adj <0.05 ###\n",
    "### Reorder and get the list of Differentially Expressed Genes (DEGs) p-values and adjusted p-values\n",
    "\n",
    "resOrdered_foetal_ctrl_vs_24h48h <- res_foetal_ctrl_vs_24h48h[order(res_foetal_ctrl_vs_24h48h$padj),]\n",
    "\n",
    "### Retain DEG with padj < 0.05 and exporting as .csv file\n",
    "resSig_foetal_ctrl_vs_24h48h <- subset(resOrdered_foetal_ctrl_vs_24h48h, padj < 0.05)\n",
    "head(resSig_foetal_ctrl_vs_24h48h)\n",
    "\n",
    "write.csv(as.data.frame(resSig_foetal_ctrl_vs_24h48h), file=\"DEGs_foetal_ctrl_vs_24h48h.csv\")\n",
    "\n",
    "\n",
    "####Upreg sig DEGs ####\n",
    "sig_DEGs_foetal_upreg_24h48h <- subset(resSig_foetal_ctrl_vs_24h48h, log2FoldChange >= 0.5)\n",
    "summary(sig_DEGs_foetal_upreg_24h48h)\n",
    "\n",
    "write.csv(as.data.frame(sig_DEGs_foetal_upreg_24h48h), file=\"sig_upreg_DEGs_foetal_24h48h.csv\")\n",
    "\n",
    "#### Downreg sig DEGs ####\n",
    "sig_DEGs_foetal_downreg_24h48h <- subset(resSig_foetal_ctrl_vs_24h48h, log2FoldChange <= -0.5)\n",
    "summary(sig_DEGs_foetal_downreg_24h48h)\n",
    "\n",
    "write.csv(as.data.frame(sig_DEGs_foetal_downreg_24h48h), file=\"sig_downreg_DEGs_foetal_24h48h.csv\")\n",
    "\n",
    "\n",
    "########################################\n",
    "\n",
    "#######Volcano plot#####\n",
    "##very basic plot##\n",
    "\n",
    "library(EnhancedVolcano)\n",
    "library(tidyverse)\n",
    "\n",
    "setwd('C:/Users/issyl/OneDrive - University of Cambridge/Research Project/DGE analysis AAI/DESeq2 analysis/RNA-Seq-AAI-analysis/results for volcano plots')\n",
    "\n",
    "read.csv(\"results_foetal_untreated_vs_treated.csv\") -> foetal_res\n",
    "\n",
    "read.csv(\"paed_res_control_vs_24+48+96.csv\") -> paed_res\n",
    "\n",
    "paed_res$X -> row.names(paed_res)\n",
    "paed_res[,-1] -> paed_res\n",
    "\n",
    "foetal_res$X -> row.names(foetal_res)\n",
    "foetal_res[,-1] -> foetal_res\n",
    "\n",
    "\n",
    "#### Paediatric volcano plot ####\n",
    "\n",
    "keyvals <- ifelse(\n",
    "  paed_res$log2FoldChange > 0.5, \"red\", \n",
    "  ifelse(paed_res$log2FoldChange < -0.5, \"royalblue\",\n",
    "         \"black\"))\n",
    "  \n",
    "keyvals[is.na(keyvals)] <- \"black\"\n",
    "names(keyvals)[keyvals == \"red\"] <- \"upregulated\"\n",
    "names(keyvals)[keyvals == \"royalblue\"] <- \"downregulated\"\n",
    "\n",
    "\n",
    "\n",
    "paed_vplot <- EnhancedVolcano(paed_res, lab=paed_res$symbol, x=\"log2FoldChange\", y=\"padj\", xlim = c(-6, 10), ylim = c(0, 25), title= \"Paediatric Untreated vs AAI-Treated DEGs\", subtitle= NULL, pCutoff= 10e-2, FCcutoff= 0.5, labSize= 3.0, \n",
    "                colCustom = keyvals, colAlpha=0.5, cutoffLineType = \"dashed\", cutoffLineCol = \"black\", cutoffLineWidth = 0.8, \n",
    "                legendPosition = \"right\", legendLabSize = 12, legendIconSize = 4, selectLab = c(\"MDM2\", \"BAX\", \"DDB2\", \"CCND1\", \"GADD45A\", \"CDKN1A\", \"XPC\", \"WWTR1\", \"FAS\", \"TP53I3\"), gridlines.major=TRUE,\n",
    "                gridlines.minor=FALSE, caption= NULL, titleLabSize= 12, subtitleLabSize= 10, ylab= \"-log10(padj)\", axisLabSize=8)\n",
    "                                                                                                                      \n",
    "\n",
    "\n",
    "\n",
    "paed_vplot + ggplot2:: coord_cartesian(xlim=c(-6, 10)) +\n",
    "                                         ggplot2::scale_x_continuous(\n",
    "                                           breaks=seq(-6,10, 2))\n",
    "                                         \n",
    "                                  \n",
    "\n",
    "#### Foetal volcano plot ####\n",
    "\n",
    "EnhancedVolcano(foetal_res, lab=foetal_res$symbol, x=\"log2FoldChange\", y=\"padj\", xlim= c(-4.75, 6), ylim = c(0, 1), title= \"Foetal Untreated vs AAI-Treated DEGs\", pCutoff= 10e-2, FCcutoff= 1, labSize= 5.0, \n",
    "                col=c(\"black\", \"green4\", \"black\", \"red3\", colAlpha=0.8), cutoffLineType = \"dashed\", cutoffLineCol = \"black\", cutoffLineWidth = 0.8, \n",
    "                legendPosition = \"right\", legendLabSize = 10, legendIconSize = 5)\n",
    "\n",
    "######################################################################################\n",
    "#### Enrichment Plots #######\n",
    "\n",
    "###### Plots to illustrate functional enrichment analysis of top differentially expressed genes ####\n",
    "\n",
    "library(tidyverse)\n",
    "library(ggplot2)\n",
    "\n",
    "setwd('C:/Users/issyl/OneDrive - University of Cambridge/Research Project/DGE analysis AAI/DESeq2 analysis/RNA-Seq-AAI-analysis/sig DEG lists for enrichment plots/enrichment plots')\n",
    "read.csv(\"metadata_paed_enrichment_plot.csv\", header = TRUE) -> enrichment_paed\n",
    "\n",
    "as.factor(enrichment_paed$pathway) -> enrichment_paed$pathway \n",
    "as.factor(enrichment_paed$fold.change) -> enrichment_paed$fold.change\n",
    "\n",
    "enrichment_paed$fold.change <- factor(enrichment_paed$fold.change, levels=c(\"Top 10 downregulated\", \"Top 10 upregulated\"))\n",
    "\n",
    "enrichment_paed$pathway <- factor(enrichment_paed$pathway, levels=unique(enrichment_paed$pathway))\n",
    "\n",
    "\n",
    "ggplot(enrichment_paed, aes(fill=fold.change)) + geom_bar(stat=\"identity\", alpha=0.6, aes(y=pathway, x=fold.enrichment)) +\n",
    "  labs(x= \"Fold Enrichment\", y= NULL, colour= NULL) +ggtitle(\"Biological Processes Enrichment in Paediatric Untreated compared to AAI-Treated Organoids\") +\n",
    "  theme(plot.title=element_text(size=12)) +\n",
    "  theme(plot.title=element_text(hjust=0.8)) + theme(legend.title=element_blank()) +\n",
    "  scale_fill_manual(values= c(\"Top 10 upregulated\" = \"red2\", \"Top 10 downregulated\" = \"royalblue\")) +\n",
    "  theme(panel.grid.major = element_blank()) + theme(panel.grid.minor = element_blank()) +\n",
    "  scale_x_continuous(limits= c(0, 10), breaks= seq(0, 10, by= 4)) + \n",
    "  theme(plot.title = element_text(margin = margin(b=20))) +\n",
    "  theme(axis.text.y = element_text(size=11)) +\n",
    "  theme(legend.text = element_text(size=10)) +\n",
    "  theme(plot.title = element_text(face = \"bold\"))\n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R 4.2.2",
   "language": "R",
   "name": "ir422"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
