{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4d9831d0",
   "metadata": {},
   "source": [
    "# # DESeq2 with ~group as design factor, included whole dataset \n",
    "## Use contrast () function to make desired control vs timepoint comparisons "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "788e5e8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Installing and loading the required packages at once\n",
    "\n",
    "if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "  install.packages(\"BiocManager\")\n",
    "\n",
    "#BiocManager::install(\"DESeq2\")\n",
    "require(DESeq2)\n",
    "#BiocManager::install(\"org.Hs.eg.db\")\n",
    "require(org.Hs.eg.db)\n",
    "#BiocManager::install(\"EnhancedVolcano\")\n",
    "require(EnhancedVolcano)\n",
    "#install.packages(\"RColorBrewer\")\n",
    "\n",
    "require(RColorBrewer)\n",
    "#install.packages(\"ggplot2\")\n",
    "require(ggplot2)\n",
    "#BiocManager::install(\"Glimma\")\n",
    "require(Glimma)\n",
    "#BiocManager::install(\"ComplexHeatmap\")\n",
    "require(ComplexHeatmap)\n",
    "#install.packages(\"gprofiler2\")\n",
    "library(gprofiler2)\n",
    "#BiocManager::install(\"tidyverse\")\n",
    "library(tidyverse)\n",
    "#BiocManager::install(\"tradeSeq\", lib= \"C:/Users/cb998/R-4.1.2/library\")\n",
    "require(tradeSeq)\n",
    "library(dplyr)\n",
    "library(data.table)\n",
    "library(tidyverse)\n",
    "require(pander)\n",
    "library(plyr)\n",
    "library(readr) \n",
    "library(AnnotationDbi)\n",
    "library(org.Hs.eg.db)\n",
    "library(EnhancedVolcano)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5cc2bd98",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set working directory ####\n",
    "setwd('C:/Users/issyl/OneDrive - University of Cambridge/Research Project/DGE analysis AAI/DESeq2 analysis/RNA-Seq-AAI-analysis/00_raw data')\n",
    "\n",
    "## Loading counts matrix and metadata\n",
    "cnts <- read.delim(file.choose(), row.names = 1)\n",
    "coldata<- read.delim(file.choose(), row.names=1)\n",
    "\n",
    "### Use this for patients study samples\n",
    "coldata <- coldata[,c(\"treatment\", \"timepoint\", \"stage\", \"ID\",\"line\",\"group\")]\n",
    "\n",
    "# Make factors \n",
    "coldata$treatment <- factor(coldata$treatment)\n",
    "coldata$timepoint <- factor(coldata$timepoint)\n",
    "coldata$stage <- factor(coldata$stage)\n",
    "coldata$ID <- factor(coldata$ID)\n",
    "coldata$line <- factor(coldata$line)\n",
    "coldata$group <- factor(coldata$group)\n",
    "\n",
    "### Make a DESeq2 design \n",
    "dds <- DESeqDataSetFromMatrix(countData = cnts,\n",
    "                              colData = coldata,\n",
    "                              design = ~ group)\n",
    "\n",
    "#### Set factor level \n",
    "dds$group <- relevel(dds$group, ref=\"fetal.control\")\n",
    "\n",
    "#### Run DESeq2 and obtain results table ####\n",
    "\n",
    "dds <- DESeq(dds)\n",
    "res <- results(dds)\n",
    "head(res)\n",
    "\n",
    "#### Remove version number from Ensembl geneid\n",
    "tmp=gsub(\"\\\\..*\",\"\",row.names(res))\n",
    "rownames(res) <- tmp\n",
    "\n",
    "#Add gene symbol\n",
    "\n",
    "res$symbol <-mapIds(org.Hs.eg.db, keys=rownames(res), column=\"SYMBOL\", keytype=\"ENSEMBL\")\n",
    "head(res)\n",
    "\n",
    "# Extract normalised counts\n",
    "\n",
    "normalized_counts <- counts(dds, normalized=TRUE)\n",
    "\n",
    "tmp=gsub(\"\\\\..*\",\"\",row.names(normalized_counts))\n",
    "rownames(normalized_counts) <- tmp\n",
    "\n",
    "write.csv (normalized_counts, file=\"Normalised_counts_04-04-2023_whole_dataset.csv\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed343b35",
   "metadata": {},
   "source": [
    "# PCA plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc56209a",
   "metadata": {},
   "outputs": [],
   "source": [
    "############## PCA PLOT ############\n",
    "require(DESeq2)\n",
    "require(org.Hs.eg.db)\n",
    "require(RColorBrewer)\n",
    "require(ggplot2)\n",
    "require(Glimma)\n",
    "library(gprofiler2)\n",
    "library(tidyverse)\n",
    "require(tradeSeq)\n",
    "library(dplyr)\n",
    "library(data.table)\n",
    "library(AnnotationDbi)\n",
    "library(org.Hs.eg.db)\n",
    "library(ggrepel)\n",
    "\n",
    "\n",
    "#### Set working directory ####\n",
    "setwd('C:/Users/issyl/OneDrive - University of Cambridge/Research Project/DGE analysis AAI/DESeq2 analysis/RNA-Seq-AAI-analysis/00_raw data')\n",
    "\n",
    "## Loading counts matrix and sample metadata\n",
    "cnts <- read.delim(file.choose(), row.names=1)\n",
    "coldata<- read.delim(file.choose(), row.names=1)\n",
    "\n",
    "\n",
    "### \n",
    "coldata <- coldata[,c(\"treatment\", \"timepoint\", \"stage\",\"ID\", \"line\", \"group\")]\n",
    "\n",
    "# Make factors \n",
    "coldata$treatment <- factor(coldata$treatment)\n",
    "coldata$timepoint <- factor(coldata$timepoint)\n",
    "coldata$stage <- factor(coldata$stage)\n",
    "coldata$ID <- factor(coldata$ID)\n",
    "coldata$line <- factor(coldata$line)\n",
    "coldata$group <- factor(coldata$group)\n",
    "\n",
    "### Make a DESeq2 design \n",
    "dds <- DESeqDataSetFromMatrix(countData = cnts,\n",
    "                              colData = coldata,\n",
    "                              design = ~ group)\n",
    "\n",
    "#### Set factor level \n",
    "dds$group <- relevel(dds$group, ref=\"fetal.control\")\n",
    "\n",
    "### Filter out feature with very low counts and keep only feature with counts' sum of =>10\n",
    "keep <- rowSums(counts(dds)) >= 10\n",
    "dds <- dds[keep,]\n",
    "\n",
    "#### Run DESeq2 and obtain results table\n",
    "\n",
    "dds <- DESeq(dds)\n",
    "res <- results(dds)\n",
    "head(res)\n",
    "### Perform PCA analysis \n",
    "\n",
    "#Plot PCA\n",
    "vstCondition <- vst(dds)\n",
    "\n",
    "pcaData <- plotPCA(vstCondition, intgroup=c(\"treatment\", \"stage\", \"ID\"), returnData=TRUE)\n",
    "\n",
    "percentVar <- round(100 * attr(pcaData, \"percentVar\"))\n",
    "\n",
    "pcaData$stage <- factor(pcaData$stage, levels=c(\"fetal\", \"paediatric\"))\n",
    "pcaData$line <- factor(pcaData$line, levels=c(\"X93\", \"X96\", \"X655\", \"X647\"))\n",
    "pcaData$treatment <- factor(pcaData$treatment, levels=c(\"untreated\", \"treated\"))\n",
    "\n",
    "\n",
    "# PCA PLOT ###\n",
    "pdf(file=\"PCA PLot no ggrepel.pdf\", onefile=TRUE, height=5, width=7, useDingbats = FALSE)\n",
    "PCAplot <- ggplot(pcaData, aes(PC1, PC2, color=stage, shape=treatment, label=name, hline = 0)) +\n",
    "  geom_point(size=2) + ggtitle(\"PCA Analysis\") +\n",
    "  geom_text(hjust= -0.5, vjust=0, size=2) +\n",
    "  xlim(-50, 50) +\n",
    "  scale_color_manual(values=c(\"fetal\" = \"darkorchid1\", \"paediatric\"=\"chartreuse4\")) +\n",
    "  scale_shape_manual(values=c(\"untreated\" = 0, \"treated\" = 16)) +\n",
    "  xlab(paste0(\"PC1: \",percentVar[1],\"% variance\")) +\n",
    "  ylab(paste0(\"PC2: \",percentVar[2],\"% variance\")) + \n",
    "  coord_fixed()+ theme(aspect.ratio = 0.7)+\n",
    "  theme(plot.title = element_text(size=16, face=\"bold\"),\n",
    "        plot.subtitle = element_text(size=14),\n",
    "        axis.text.x=element_text(size=14,colour = \"black\"),\n",
    "        axis.text.y=element_text(size=14,colour = \"black\"),\n",
    "        axis.title.x = element_text(size=14),\n",
    "        axis.title.y = element_text(size=14),\n",
    "        legend.title = element_text(size=14, colour = \"black\", face = \"bold\"),\n",
    "        legend.text = element_text(size=14),\n",
    "        strip.text = element_text(size = 14, face = \"bold\"),\n",
    "        panel.grid.minor.x=element_blank(),\n",
    "        panel.grid.major.x=element_blank(),\n",
    "        panel.grid.major.y = element_blank(),\n",
    "        panel.grid.minor.y = element_blank(),\n",
    "        panel.background = element_rect(fill = \"white\"),\n",
    "        panel.border = element_rect(colour = \"black\", fill=FALSE))\n",
    "\n",
    "plot\n",
    "print(PCAplot)\n",
    "dev.off()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ac31244",
   "metadata": {},
   "source": [
    "# Obtain results for each desired contrast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9120123",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Obtain results for each desired contrast ####\n",
    "\n",
    "\n",
    "#### Fetal vs paed baseline DGE ####\n",
    "# NB Paed control is baseline here, so LFC >0 means expression is upreg in fetal compared to paed \n",
    "\n",
    "res_fetal.control_vs_paed.control <- results(dds, contrast=c(\"group\",\"fetal.control\",\"paed.control\"), alpha=0.05)\n",
    "summary(res_fetal.control_vs_paed.control)\n",
    "\n",
    "\n",
    "##### FETAL ONLY #####\n",
    "\n",
    "### Fetal control vs 6h_treated\n",
    "\n",
    "res_fetal.control_vs_fetal.6h <- results(dds, contrast=c(\"group\",\"fetal.6h\",\"fetal.control\"), alpha=0.05)\n",
    "summary(res_fetal.control_vs_fetal.6h)\n",
    "\n",
    "### Fetal control vs 24h_treated\n",
    "\n",
    "res_fetal.control_vs_fetal.24h <- results(dds, contrast=c(\"group\",\"fetal.24h\",\"fetal.control\"), alpha=0.05)\n",
    "summary(res_fetal.control_vs_fetal.24h)\n",
    "\n",
    "### Fetal control vs 48h_treated\n",
    "\n",
    "res_fetal.control_vs_fetal.48h <- results(dds, contrast=c(\"group\",\"fetal.48h\",\"fetal.control\"), alpha=0.05)\n",
    "summary(res_fetal.control_vs_fetal.48h)\n",
    "\n",
    "### Fetal control vs 96h_treated\n",
    "\n",
    "res_fetal.control_vs_fetal.96h <- results(dds, contrast=c(\"group\",\"fetal.96h\",\"fetal.control\"), alpha=0.05)\n",
    "summary(res_fetal.control_vs_fetal.96h)\n",
    "\n",
    "\n",
    "##### PAED ONLY #####\n",
    "\n",
    "### Paed control vs 6h_treated\n",
    "\n",
    "res_paed.control_vs_paed.6h <- results(dds, contrast=c(\"group\",\"paed.6h\",\"paed.control\"), alpha=0.05)\n",
    "summary(res_paed.control_vs_paed.6h)\n",
    "\n",
    "### Paed control vs 24h_treated\n",
    "\n",
    "res_paed.control_vs_paed.24h <- results(dds, contrast=c(\"group\",\"paed.24h\",\"paed.control\"), alpha=0.05)\n",
    "summary(res_paed.control_vs_paed.24h)\n",
    "\n",
    "### Paed control vs 48h_treated \n",
    "\n",
    "res_paed.control_vs_paed.48h <- results(dds, contrast=c(\"group\",\"paed.48h\",\"paed.control\"), alpha=0.05)\n",
    "summary(res_paed.control_vs_paed.48h)\n",
    "\n",
    "### Paed control vs 96h_treated#\n",
    "\n",
    "res_paed.control_vs_paed.96h <- results(dds, contrast=c(\"group\",\"paed.96h\",\"paed.control\"), alpha=0.05)\n",
    "summary(res_paed.control_vs_paed.96h)\n",
    "\n",
    "#################################################################################################\n",
    "#### Add gene symbol as a column ####\n",
    "\n",
    "tmp=gsub(\"\\\\..*\",\"\",row.names(res_fetal.control_vs_paed.control))\n",
    "rownames(res_fetal.control_vs_paed.control) <- tmp\n",
    "\n",
    "res_fetal.control_vs_paed.control$symbol <-mapIds(org.Hs.eg.db, keys=rownames(res_fetal.control_vs_paed.control), column=\"SYMBOL\", keytype=\"ENSEMBL\")\n",
    "\n",
    "\n",
    "tmp=gsub(\"\\\\..*\",\"\",row.names(res_fetal.control_vs_fetal.6h))\n",
    "rownames(res_fetal.control_vs_fetal.6h) <- tmp\n",
    "\n",
    "res_fetal.control_vs_fetal.6h$symbol <-mapIds(org.Hs.eg.db, keys=rownames(res_fetal.control_vs_fetal.6h), column=\"SYMBOL\", keytype=\"ENSEMBL\")\n",
    "\n",
    "\n",
    "tmp=gsub(\"\\\\..*\",\"\",row.names(res_fetal.control_vs_fetal.24h))\n",
    "rownames(res_fetal.control_vs_fetal.24h) <- tmp\n",
    "\n",
    "res_fetal.control_vs_fetal.24h$symbol <-mapIds(org.Hs.eg.db, keys=rownames(res_fetal.control_vs_fetal.24h), column=\"SYMBOL\", keytype=\"ENSEMBL\")\n",
    "\n",
    "\n",
    "tmp=gsub(\"\\\\..*\",\"\",row.names(res_fetal.control_vs_fetal.48h))\n",
    "rownames(res_fetal.control_vs_fetal.48h) <- tmp\n",
    "\n",
    "res_fetal.control_vs_fetal.48h$symbol <-mapIds(org.Hs.eg.db, keys=rownames(res_fetal.control_vs_fetal.48h), column=\"SYMBOL\", keytype=\"ENSEMBL\")\n",
    "\n",
    "\n",
    "tmp=gsub(\"\\\\..*\",\"\",row.names(res_fetal.control_vs_fetal.96h))\n",
    "rownames(res_fetal.control_vs_fetal.96h) <- tmp\n",
    "\n",
    "res_fetal.control_vs_fetal.96h$symbol <-mapIds(org.Hs.eg.db, keys=rownames(res_fetal.control_vs_fetal.96h), column=\"SYMBOL\", keytype=\"ENSEMBL\")\n",
    "\n",
    "\n",
    "tmp=gsub(\"\\\\..*\",\"\",row.names(res_paed.control_vs_paed.6h))\n",
    "rownames(res_paed.control_vs_paed.6h) <- tmp\n",
    "\n",
    "res_paed.control_vs_paed.6h$symbol <-mapIds(org.Hs.eg.db, keys=rownames(res_paed.control_vs_paed.6h), column=\"SYMBOL\", keytype=\"ENSEMBL\")\n",
    "\n",
    "\n",
    "\n",
    "tmp=gsub(\"\\\\..*\",\"\",row.names(res_paed.control_vs_paed.24h))\n",
    "rownames(res_paed.control_vs_paed.24h) <- tmp\n",
    "\n",
    "res_paed.control_vs_paed.24h$symbol <-mapIds(org.Hs.eg.db, keys=rownames(res_paed.control_vs_paed.24h), column=\"SYMBOL\", keytype=\"ENSEMBL\")\n",
    "\n",
    "\n",
    "\n",
    "tmp=gsub(\"\\\\..*\",\"\",row.names(res_paed.control_vs_paed.48h))\n",
    "rownames(res_paed.control_vs_paed.48h) <- tmp\n",
    "\n",
    "res_paed.control_vs_paed.48h$symbol <-mapIds(org.Hs.eg.db, keys=rownames(res_paed.control_vs_paed.48h), column=\"SYMBOL\", keytype=\"ENSEMBL\")\n",
    "\n",
    "\n",
    "tmp=gsub(\"\\\\..*\",\"\",row.names(res_paed.control_vs_paed.96h))\n",
    "rownames(res_paed.control_vs_paed.96h) <- tmp\n",
    "\n",
    "res_paed.control_vs_paed.96h$symbol <-mapIds(org.Hs.eg.db, keys=rownames(res_paed.control_vs_paed.96h), column=\"SYMBOL\", keytype=\"ENSEMBL\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48eef3e4",
   "metadata": {},
   "source": [
    "# Obtaining lists of sig DEGs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ccadefd",
   "metadata": {},
   "outputs": [],
   "source": [
    "##### Fetal control vs paed control ####\n",
    "#### Obtain just those genes with p-adj <0.05 ###\n",
    "\n",
    "### Reorder and get the list of Differentially Expressed Genes (DEGs) p-values and adjusted p-values\n",
    "\n",
    "resOrdered_fetal.control_vs_paed.control <- res_fetal.control_vs_paed.control[order(res_fetal.control_vs_paed.control$padj),]\n",
    "\n",
    "### Retain DEG with padj < 0.05 and exporting as .csv file\n",
    "resSig_fetal.control_vs_paed.control <- subset(resOrdered_fetal.control_vs_paed.control, padj < 0.05)\n",
    "head(resSig_fetal.control_vs_paed.control)\n",
    "\n",
    "write.csv(as.data.frame(resSig_fetal.control_vs_paed.control), file=\"DEGs_fetal_control_vs_paed_control.csv\")\n",
    "\n",
    "###### FETAL ONLY DEG lists ######\n",
    "##### Fetal control vs 6hr ######\n",
    "#### Obtain just those genes with p-adj <0.05 ###\n",
    "\n",
    "### Reorder and get the list of Differentially Expressed Genes (DEGs) p-values and adjusted p-values\n",
    "\n",
    "resOrdered_fetal.control_vs_fetal.6h <- res_fetal.control_vs_fetal.6h[order(res_fetal.control_vs_fetal.6h$padj),]\n",
    "\n",
    "### Retain DEG with padj < 0.05 and exporting as .csv file\n",
    "resSig_fetal.control_vs_fetal.6h <- subset(resOrdered_fetal.control_vs_fetal.6h, padj < 0.05)\n",
    "head(resSig_fetal.control_vs_fetal.6h)\n",
    "\n",
    "write.csv(as.data.frame(resSig_fetal.control_vs_fetal.6h), file=\"DEGs_fetal_control_vs_fetal_6h.csv\")\n",
    "\n",
    "\n",
    "##### Fetal control vs 24hr #####\n",
    "#### Obtain just those genes with p-adj <0.05 ###\n",
    "\n",
    "### Reorder and get the list of Differentially Expressed Genes (DEGs) p-values and adjusted p-values\n",
    "\n",
    "resOrdered_fetal.control_vs_fetal.24h <- res_fetal.control_vs_fetal.24h[order(res_fetal.control_vs_fetal.24h$padj),]\n",
    "\n",
    "### Retain DEG with padj < 0.05 and exporting as .csv file\n",
    "resSig_fetal.control_vs_fetal.24h <- subset(resOrdered_fetal.control_vs_fetal.24h, padj < 0.05)\n",
    "head(resSig_fetal.control_vs_fetal.24h)\n",
    "\n",
    "write.csv(as.data.frame(resSig_fetal.control_vs_fetal.24h), file=\"DEGs_fetal_control_vs_fetal_24h.csv\")\n",
    "\n",
    "\n",
    "##### Fetal control vs 48hr #####\n",
    "#### Obtain just those genes with p-adj <0.05 ###\n",
    "\n",
    "### Reorder and get the list of Differentially Expressed Genes (DEGs) p-values and adjusted p-values\n",
    "\n",
    "resOrdered_fetal.control_vs_fetal.48h <- res_fetal.control_vs_fetal.48h[order(res_fetal.control_vs_fetal.48h$padj),]\n",
    "\n",
    "### Retain DEG with padj < 0.05 and exporting as .csv file\n",
    "resSig_fetal.control_vs_fetal.48h <- subset(resOrdered_fetal.control_vs_fetal.48h, padj < 0.05)\n",
    "head(resSig_fetal.control_vs_fetal.48h)\n",
    "\n",
    "write.csv(as.data.frame(resSig_fetal.control_vs_fetal.48h), file=\"DEGs_fetal_control_vs_fetal_48h.csv\")\n",
    "\n",
    "\n",
    "##### Fetal control vs 96hr #####\n",
    "#### Obtain just those genes with p-adj <0.05 ###\n",
    "\n",
    "### Reorder and get the list of Differentially Expressed Genes (DEGs) p-values and adjusted p-values\n",
    "\n",
    "resOrdered_fetal.control_vs_fetal.96h <- res_fetal.control_vs_fetal.96h[order(res_fetal.control_vs_fetal.96h$padj),]\n",
    "\n",
    "### Retain DEG with padj < 0.05 and exporting as .csv file\n",
    "resSig_fetal.control_vs_fetal.96h <- subset(resOrdered_fetal.control_vs_fetal.96h, padj < 0.05)\n",
    "head(resSig_fetal.control_vs_fetal.96h)\n",
    "\n",
    "write.csv(as.data.frame(resSig_fetal.control_vs_fetal.96h), file=\"DEGs_fetal_control_vs_fetal_96h.csv\")\n",
    "\n",
    "####### PAED ONLY DEG LISTS ########\n",
    "\n",
    "##### Paed control vs 6hr ######\n",
    "#### Obtain just those genes with p-adj <0.05 ###\n",
    "\n",
    "### Reorder and get the list of Differentially Expressed Genes (DEGs) p-values and adjusted p-values\n",
    "\n",
    "resOrdered_paed.control_vs_paed.6h <- res_paed.control_vs_paed.6h[order(res_paed.control_vs_paed.6h$padj),]\n",
    "\n",
    "### Retain DEG with padj < 0.05 and exporting as .csv file\n",
    "resSig_paed.control_vs_paed.6h <- subset(resOrdered_paed.control_vs_paed.6h, padj < 0.05)\n",
    "head(resSig_paed.control_vs_paed.6h)\n",
    "\n",
    "write.csv(as.data.frame(resSig_paed.control_vs_paed.6h), file=\"DEGs_paed_control_vs_paed_6h.csv\")\n",
    "\n",
    "\n",
    "##### Paed control vs 24hr #####\n",
    "#### Obtain just those genes with p-adj <0.05 ###\n",
    "\n",
    "### Reorder and get the list of Differentially Expressed Genes (DEGs) p-values and adjusted p-values\n",
    "\n",
    "resOrdered_paed.control_vs_paed.24h <- res_paed.control_vs_paed.24h[order(res_paed.control_vs_paed.24h$padj),]\n",
    "\n",
    "### Retain DEG with padj < 0.05 and exporting as .csv file\n",
    "resSig_paed.control_vs_paed.24h <- subset(resOrdered_paed.control_vs_paed.24h, padj < 0.05)\n",
    "head(resSig_paed.control_vs_paed.24h)\n",
    "\n",
    "write.csv(as.data.frame(resSig_paed.control_vs_paed.24h), file=\"DEGs_paed_control_vs_paed_24h.csv\")\n",
    "\n",
    "\n",
    "###### Paed control vs 48h #######\n",
    "#### Obtain just those genes with p-adj <0.05 ###\n",
    "\n",
    "### Reorder and get the list of Differentially Expressed Genes (DEGs) p-values and adjusted p-values\n",
    "\n",
    "resOrdered_paed.control_vs_paed.48h <- res_paed.control_vs_paed.48h[order(res_paed.control_vs_paed.48h$padj),]\n",
    "\n",
    "### Retain DEG with padj < 0.05 and exporting as .csv file\n",
    "resSig_paed.control_vs_paed.48h <- subset(resOrdered_paed.control_vs_paed.48h, padj < 0.05)\n",
    "head(resSig_paed.control_vs_paed.48h)\n",
    "\n",
    "write.csv(as.data.frame(resSig_paed.control_vs_paed.48h), file=\"DEGs_paed_control_vs_paed_48h.csv\")\n",
    "\n",
    "\n",
    "###### Paed control vs 96hr ######\n",
    "#### Obtain just those genes with p-adj <0.05 ###\n",
    "\n",
    "### Reorder and get the list of Differentially Expressed Genes (DEGs) p-values and adjusted p-values\n",
    "\n",
    "resOrdered_paed.control_vs_paed.96h <- res_paed.control_vs_paed.96h[order(res_paed.control_vs_paed.96h$padj),]\n",
    "\n",
    "### Retain DEG with padj < 0.05 and exporting as .csv file\n",
    "resSig_paed.control_vs_paed.96h <- subset(resOrdered_paed.control_vs_paed.96h, padj < 0.05)\n",
    "head(resSig_paed.control_vs_paed.96h)\n",
    "\n",
    "write.csv(as.data.frame(resSig_paed.control_vs_paed.96h), file=\"DEGs_paed_control_vs_paed_96h.csv\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a459e26a",
   "metadata": {},
   "source": [
    "# # LOOKING FOR JOINT AND UNIQUE HITS BETWEEN FETAL AND PAEDIATRIC SIG DEGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0535553b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Joint and Unqiue hits ##\n",
    "\n",
    "### 6hr timepoint ###\n",
    "\n",
    "# Make a shorter name\n",
    "resSig_fetal.control_vs_fetal.6h -> DEGs_fetal_6h\n",
    "resSig_paed.control_vs_paed.6h -> DEGs_paed_6h\n",
    "\n",
    "# Make geneid a 'proper' column in DEGs table\n",
    "DEGs_fetal_6h$geneid <- rownames(DEGs_fetal_6h)\n",
    "DEGs_paed_6h$geneid <- rownames(DEGs_paed_6h)\n",
    "\n",
    "# Looking for number of joint hits between fetal and paed samples out of DEGs identified in control vs 6hr DGE analyses\n",
    "\n",
    "overlap <- subset(DEGs_fetal_6h, geneid %in% DEGs_paed_6h$geneid)\n",
    "nrow(overlap)\n",
    "\n",
    "# Looking for number of unique hits in fetal samples (i.e. gene hits that are not in the paed samples\n",
    "\n",
    "missing.1 <- subset(DEGs_fetal_6h, !(geneid %in% DEGs_paed_6h$geneid))\n",
    "nrow(missing.1)\n",
    "\n",
    "# Looking for number of unique hits in paed samples (i.e. gene hits that are not in fetal samples)\n",
    "\n",
    "missing.2 <- subset(DEGs_paed_6h, !(geneid %in% DEGs_fetal_6h$geneid))\n",
    "\n",
    "### Export the hits unique to fetal samples ###\n",
    "\n",
    "write.csv(as.data.frame(missing.1), file=\"DEGs_6h_unique_to_fetal.csv\")\n",
    "\n",
    "## Export joint hits at 6hr\n",
    "\n",
    "write.csv(as.data.frame(overlap), file=\"DEGs_6h_joint_hits.csv\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "### 24hr timepoint ###\n",
    "\n",
    "# Make a shorter name\n",
    "resSig_fetal.control_vs_fetal.24h -> DEGs_fetal_24h\n",
    "resSig_paed.control_vs_paed.24h -> DEGs_paed_24h\n",
    "\n",
    "# Make geneid a 'proper' column in DEGs table\n",
    "DEGs_fetal_24h$geneid <- rownames(DEGs_fetal_24h)\n",
    "DEGs_paed_24h$geneid <- rownames(DEGs_paed_24h)\n",
    "\n",
    "# Looking for number of joint hits between fetal and paed samples out of DEGs identified in control vs 6hr DGE analyses\n",
    "\n",
    "overlap <- subset(DEGs_fetal_24h, geneid %in% DEGs_paed_24h$geneid)\n",
    "nrow(overlap)\n",
    "\n",
    "# Looking for number of unique hits in fetal samples (i.e. gene hits that are not in the paed samples\n",
    "\n",
    "missing.1 <- subset(DEGs_fetal_24h, !(geneid %in% DEGs_paed_24h$geneid))\n",
    "nrow(missing.1)\n",
    "\n",
    "# Looking for number of unique hits in paed samples (i.e. gene hits that are not in fetal samples)\n",
    "\n",
    "missing.2 <- subset(DEGs_paed_24h, !(geneid %in% DEGs_fetal_24h$geneid))\n",
    "\n",
    "### Export the hits unique to fetal samples ###\n",
    "\n",
    "write.csv(as.data.frame(missing.1), file=\"DEGs_24h_unique_to_fetal.csv\")\n",
    "\n",
    "## Export joint hits at 24hr\n",
    "\n",
    "write.csv(as.data.frame(overlap), file=\"DEGs_24h_joint_hits.csv\")\n",
    "\n",
    "## Export hits unique to paed samples \n",
    "\n",
    "write.csv(as.data.frame(missing.2), file= \"DEGs_24h_unique_to_paed.csv\")\n",
    "\n",
    "\n",
    "\n",
    "#### 48hr timepoint ####\n",
    "\n",
    "# Make a shorter name\n",
    "resSig_fetal.control_vs_fetal.48h -> DEGs_fetal_48h\n",
    "resSig_paed.control_vs_paed.48h -> DEGs_paed_48h\n",
    "\n",
    "# Make geneid a 'proper' column in DEGs table\n",
    "DEGs_fetal_48h$geneid <- rownames(DEGs_fetal_48h)\n",
    "DEGs_paed_48h$geneid <- rownames(DEGs_paed_48h)\n",
    "\n",
    "# Looking for number of joint hits between fetal and paed samples out of DEGs identified in control vs 6hr DGE analyses\n",
    "\n",
    "overlap <- subset(DEGs_fetal_48h, geneid %in% DEGs_paed_48h$geneid)\n",
    "nrow(overlap)\n",
    "\n",
    "# Looking for number of unique hits in fetal samples (i.e. gene hits that are not in the paed samples\n",
    "\n",
    "missing.1 <- subset(DEGs_fetal_48h, !(geneid %in% DEGs_paed_48h$geneid))\n",
    "nrow(missing.1)\n",
    "\n",
    "# Looking for number of unique hits in paed samples (i.e. gene hits that are not in fetal samples)\n",
    "\n",
    "missing.2 <- subset(DEGs_paed_48h, !(geneid %in% DEGs_fetal_48h$geneid))\n",
    "\n",
    "### Export the hits unique to fetal samples ###\n",
    "\n",
    "write.csv(as.data.frame(missing.1), file=\"DEGs_48h_unique_to_fetal.csv\")\n",
    "\n",
    "## Export joint hits at 48hr\n",
    "\n",
    "write.csv(as.data.frame(overlap), file=\"DEGs_48h_joint_hits.csv\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#### 96hr timepoint ####\n",
    "\n",
    "# Make a shorter name\n",
    "resSig_fetal.control_vs_fetal.96h -> DEGs_fetal_96h\n",
    "resSig_paed.control_vs_paed.96h -> DEGs_paed_96h\n",
    "\n",
    "# Make geneid a 'proper' column in DEGs table\n",
    "DEGs_fetal_96h$geneid <- rownames(DEGs_fetal_96h)\n",
    "DEGs_paed_96h$geneid <- rownames(DEGs_paed_96h)\n",
    "\n",
    "# Looking for number of joint hits between fetal and paed samples out of DEGs identified in control vs 96hr DGE analyses\n",
    "\n",
    "overlap <- subset(DEGs_fetal_96h, geneid %in% DEGs_paed_96h$geneid)\n",
    "nrow(overlap)\n",
    "\n",
    "# Looking for number of unique hits in fetal samples (i.e. gene hits that are not in the paed samples\n",
    "\n",
    "missing.1 <- subset(DEGs_fetal_96h, !(geneid %in% DEGs_paed_96h$geneid))\n",
    "nrow(missing.1)\n",
    "\n",
    "# Looking for number of unique hits in paed samples (i.e. gene hits that are not in fetal samples)\n",
    "\n",
    "missing.2 <- subset(DEGs_paed_96h, !(geneid %in% DEGs_fetal_96h$geneid))\n",
    "\n",
    "### Export the hits unique to fetal samples ###\n",
    "\n",
    "write.csv(as.data.frame(missing.1), file=\"DEGs_96h_unique_to_fetal.csv\")\n",
    "\n",
    "## Export joint hits at 48hr\n",
    "\n",
    "write.csv(as.data.frame(overlap), file=\"DEGs_96h_joint_hits.csv\")\n",
    "\n",
    "## Export hits unique to paed samples at 96hr timepoint \n",
    "write.csv(as.data.frame(missing.2), file=\"DEGs_96h_unique_to_paed.csv\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f7dfa80",
   "metadata": {},
   "source": [
    "# Simple ggplot RNA-Seq normalised counts across timeseries \n",
    "### Control and treated, Fetal and Paediatric "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "117b7ae2",
   "metadata": {},
   "outputs": [],
   "source": [
    "######### SIMPLE EXPLORATORY RNA-SEQ NORM COUNTS PLOTS OVER TIME ########\n",
    "##### Load packages####\n",
    "library(ggpubr)\n",
    "install.packages(\"tidyverse\")\n",
    "library(tidyverse)\n",
    "library(dplyr)\n",
    "library(ggplot2)\n",
    "library(plyr)\n",
    "## Setting the working directory (add the path to where you store your normalised counts and simples information)\n",
    "setwd(\"C:/Users/issyl/OneDrive - University of Cambridge/Research Project/DGE analysis AAI/DESeq2 analysis/RNA-Seq-AAI-analysis/00_raw data\")\n",
    "getwd()\n",
    "\n",
    "## Importing the normalised counts and samples information\n",
    "#Let's assume you have a .csv file format\n",
    "\n",
    "counts <- read.csv(file.choose())\n",
    "head(counts)\n",
    "meta <- read.table(file.choose(), header=T, sep=\"\\t\", stringsAsFactors=F)\n",
    "head(meta)\n",
    "\n",
    "# add the rownames as a proper column\n",
    "#Let's set the ensembl ID as colnames for us to easily select/change features later\n",
    "\n",
    "colnames(counts)[1] <- \"feature\"\n",
    "head(counts)\n",
    "\n",
    "# Select your feature of interest/gene\n",
    "### p21 ###\n",
    "selected_p21 <- c(\"ENSG00000124762\")\n",
    "\n",
    "counts_selected <- counts[counts$feature %in% selected_p21, ] \n",
    "counts_selected \n",
    "\n",
    "# Generate a long format then combine the counts table and the sample information \n",
    "counts_long <- counts_selected  %>% gather(sample, count, 2:ncol(counts_selected ), factor_key = TRUE)\n",
    "counts_long\n",
    "\n",
    "\n",
    "# Merge the two tables together\n",
    "counts_long_annot <- merge(counts_long, meta, by = \"sample\")\n",
    "counts_long_annot\n",
    "\n",
    "\n",
    "# Arrange data based on a factor you want (e.g stage: fetal vs adult) then enter the levels in the order you want (e.g fetal then adult) \n",
    "#### STAGE ####\n",
    "counts_long_annot$stage <- factor(counts_long_annot$stage, levels = c(\"fetal\", \"adult\"))\n",
    "counts_long_annot\n",
    "\n",
    "#### GROUP ####\n",
    "counts_long_annot$group <- factor(counts_long_annot$group, levels = c(\"fetal.control\", \"fetal.6h\", \"fetal.24h\", \"fetal.48h\", \"fetal.96h\", \"paed.control\", \"paed.6h\", \"paed.24h\", \"paed.48h\", \"paed.96h\"))\n",
    "counts_long_annot\n",
    "\n",
    "#### LINE####\n",
    "counts_long_annot$line <- factor(counts_long_annot$line, levels = c(\"X93\", \"X96\", \"X647\", \"X655\"))\n",
    "counts_long_annot\n",
    "\n",
    "#### TIMEPOINT ####\n",
    "counts_long_annot$timepoint <- factor(counts_long_annot$timepoint, levels = c(\"0C\", \"48C\", \"6\", \"24\", \"48\", \"96\"))\n",
    "counts_long_annot\n",
    "\n",
    "\n",
    "# Changing name to a short one\n",
    "data <- counts_long_annot\n",
    "data\n",
    "\n",
    "\n",
    "## Use ggplot to plot the time series data I assume you will use 0 hours, 6hours, 24 hours, 48hours + 48h untreated then recovery\n",
    "\n",
    "ggplot(data = data, aes(x=timepoint, y=count, group=line, colour=stage)) + geom_line() + labs(y= \"Norm Counts\", x= \"Timepoint\") + ggtitle(\"p21 Expression levels timeseries\")\n",
    "\n",
    "\n",
    "### ggplot with 'shape' indicating the biological replicate and colour indicating stage ##\n",
    "gg <- ggplot(data = data) \n",
    "gg <- gg + geom_line(aes(x=timepoint, y=count, group=line, colour=stage))\n",
    "gg <- gg + geom_point(aes(x=timepoint, y=count, shape=line), size=1.5)\n",
    "gg <- gg + labs(y= \"Norm Counts\", x= \"Timepoint\")\n",
    "gg <- gg + ggtitle(\"WNT7A: Timecourse of Normalised Counts\")\n",
    "gg\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d986dd76",
   "metadata": {},
   "source": [
    "# Heatmap script\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddcc6376",
   "metadata": {},
   "outputs": [],
   "source": [
    "##### HEATMAP SCRIPT ####\n",
    "\n",
    "#### Load packages ####\n",
    "#install.packages(\"pheatmap\")\n",
    "require(pheatmap)\n",
    "#BiocManager::install(\"RColorBrewer\")\n",
    "library(RColorBrewer)\n",
    "library(tidyverse)\n",
    "library(dplyr)\n",
    "library(pheatmap)\n",
    "\n",
    "install.packages(\"colorRampPalette\")\n",
    "install.packages(\"Rtools\")\n",
    "library(AnnotationDbi)\n",
    "library(org.Hs.eg.db)\n",
    "\n",
    "### Set working directory ###\n",
    "\n",
    "setwd('C:/Users/issyl/OneDrive - University of Cambridge/Research Project/DGE analysis AAI/DESeq2 analysis/RNA-Seq-AAI-analysis/normalised counts/Norm counts whole dataset included')\n",
    "\n",
    "### Read in normalised counts data ###\n",
    "\n",
    "read.csv(\"USE_THIS_Normalised_counts_whole_dataset_low_counts_removed_5-4-23.csv\") -> norm_counts \n",
    "\n",
    "## Make geneid the rownames ##\n",
    "norm_counts$geneid -> rownames(norm_counts)\n",
    "head(norm_counts)\n",
    "\n",
    "### To make just one geneid column ##\n",
    "norm_counts <- norm_counts[, -1]\n",
    "head(norm_counts)\n",
    "\n",
    "### Add gene symbols ###\n",
    "norm_counts$symbol <-mapIds(org.Hs.eg.db, keys=rownames(norm_counts), column=\"SYMBOL\", keytype=\"ENSEMBL\")\n",
    "\n",
    "### Select GOIs you want to show in the heatmap from the normalised counts ###\n",
    "norm_counts[c('ENSG00000141510', 'ENSG00000135679', 'ENSG00000116717', 'ENSG00000134574', 'ENSG00000087088', 'ENSG00000124762', 'ENSG00000110092' ),] -> p53_GOIs\n",
    "\n",
    "## Make the gene symbols the row names ##\n",
    "p53_GOIs$symbol -> rownames(p53_GOIs)\n",
    "\n",
    "## Remove other gene symbol column ##\n",
    "p53_GOIs <- p53_GOIs[, -25]\n",
    "head(p53_GOIs)\n",
    "\n",
    "## Reorder samples to desired order for heatmap ##\n",
    "p53_GOIs <- p53_GOIs[, c(1, 2, 7, 8, 3, 9, 4, 10, 5, 11, 6, 12, 13, 14, 19, 20, 15, 21, 16, 22, 17, 23, 18, 24)]\n",
    "head(p53_GOIs)\n",
    "\n",
    "### Make it a matrix ###\n",
    "as.matrix(p53_GOIs) -> p53_GOIs\n",
    "\n",
    "#### Need an annotation file too ###\n",
    "setwd(\"C:/Users/issyl/OneDrive - University of Cambridge/Research Project/DGE analysis AAI/DESeq2 analysis/RNA-Seq-AAI-analysis/00_raw data\")\n",
    "read.delim(\"metadataRNAseqAAI.txt\") -> anno\n",
    "rownames(anno) <- anno$sample\n",
    "anno$sample <- NULL\n",
    "anno[, c(1, 2, 3)] -> anno\n",
    "\n",
    "### Make annotation file order match the counts data order ###\n",
    "anno <- anno[c(1, 2, 7, 8, 3, 9, 4, 10, 5, 11, 6, 12, 13, 14, 19, 20, 15, 21, 16, 22, 17, 23, 18, 24),]\n",
    "\n",
    "\n",
    "#### HEATMAP ###\n",
    "cols <- list(treatment=c(untreated=\"grey80\", treated= \"cornflowerblue\"), stage=c(fetal=\"grey60\", adult=\"aquamarine4\"), timepoint=c())\n",
    "\n",
    "heat <- pheatmap(p53_GOIs, color=colorRampPalette(c(\"blue\",\"white\",\"red3\"))(50), clustering_distance_cols=\"euclidean\", clustering_method=\"complete\", cluster_rows=F, cluster_cols=F, treeheight_row=10, treeheight_col=50, cellheight=15, cellwidth=15, border_color=F, scale=\"row\", show_colnames=T, show_rownames=T, annotation_col=anno, annotation_colors=cols)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "930d7922",
   "metadata": {},
   "source": [
    "# Heatmap script if want to have a gene list as an input\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09670f39",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Heatmap script using a gene list to select genes of interest regarding a particular pathway/GO term you're interested in ####\n",
    "\n",
    "require(pheatmap)\n",
    "library(RColorBrewer)\n",
    "library(tidyverse)\n",
    "library(dplyr)\n",
    "library(pheatmap)\n",
    "library(AnnotationDbi)\n",
    "library(org.Hs.eg.db)\n",
    "\n",
    "### Set working directory ###\n",
    "\n",
    "setwd('C:/Users/issyl/OneDrive - University of Cambridge/Research Project/DGE analysis AAI/DESeq2 analysis/RNA-Seq-AAI-analysis/normalised counts/Norm counts whole dataset included')\n",
    "\n",
    "### Read in normalised counts data ###\n",
    "read.csv(\"USE_THIS_Normalised_counts_whole_dataset_low_counts_removed_5-4-23.csv\") -> list2\n",
    "\n",
    "##### List1 is the list of preselected GOIs to plot on heatmap, list2 is normalised counts dataset #####\n",
    "setwd('C:/Users/issyl/OneDrive - University of Cambridge/Research Project/DGE analysis AAI/DESeq2 analysis/RNA-Seq-AAI-analysis/normalised counts/Norm counts whole dataset included/gene lists')\n",
    "\n",
    "read.csv(\"NER list.csv\") -> list1\n",
    "\n",
    "heatmap_data <-merge(x= list1,y= list2, by= \"geneid\", all.x = TRUE)\n",
    "\n",
    "## Make geneid the rownames ##\n",
    "heatmap_data$geneid -> rownames(heatmap_data)\n",
    "head(heatmap_data)\n",
    "\n",
    "### To make just one geneid column ##\n",
    "heatmap_data <- heatmap_data[, -1]\n",
    "head(heatmap_data)\n",
    "\n",
    "### Add gene symbols ###\n",
    "heatmap_data$symbol <-mapIds(org.Hs.eg.db, keys=rownames(heatmap_data), column=\"SYMBOL\", keytype=\"ENSEMBL\")\n",
    "\n",
    "## Make the gene symbols the row names ##\n",
    "heatmap_data$symbol -> rownames(heatmap_data)\n",
    "\n",
    "#### Remove unwanted columns ###\n",
    "## Remove other gene symbol column ##\n",
    "heatmap_data <- heatmap_data[, -25]\n",
    "\n",
    "\n",
    "## Reorder samples to desired order for heatmap ##\n",
    "heatmap_data <- heatmap_data[, c(1, 2, 7, 8, 3, 9, 4, 10, 5, 11, 6, 12, 13, 14, 19, 20, 15, 21, 16, 22, 17, 23, 18, 24)]\n",
    "heatmap_data <- heatmap_data[c(7, 8, 6, 3, 2, 1, 4, 5, 12, 11, 10, 9),]\n",
    "head(heatmap_data)\n",
    "\n",
    "### Make it a matrix ###\n",
    "as.matrix(heatmap_data) -> heatmap_data\n",
    "\n",
    "#### Need an annotation file too, describing samples, eg foetal and paediatric, treated and untreated###\n",
    "setwd(\"C:/Users/issyl/OneDrive - University of Cambridge/Research Project/DGE analysis AAI/DESeq2 analysis/RNA-Seq-AAI-analysis/00_raw data\")\n",
    "read.delim(\"metadataRNAseqAAI2.txt\") -> anno\n",
    "rownames(anno) <- anno$sample\n",
    "anno$sample <- NULL\n",
    "anno[, c(1, 3)] -> anno\n",
    "\n",
    "### Make annotation file order match the counts data order ###\n",
    "anno <- anno[c(1, 2, 7, 8, 3, 9, 4, 10, 5, 11, 6, 12, 13, 14, 19, 20, 15, 21, 16, 22, 17, 23, 18, 24),]\n",
    "\n",
    "\n",
    "#### HEATMAP ###\n",
    "cols <- list(treatment=c(untreated=\"grey20\", treated= \"cornflowerblue\"), stage=c(fetal=\"grey60\", paediatric=\"aquamarine4\"))\n",
    "\n",
    "pdf(file=\"NER heatmap.pdf\", onefile=TRUE, height=15, width=10, useDingbats = FALSE)\n",
    "heat <- pheatmap(heatmap_data, color=colorRampPalette(c(\"#2900f7\",\"#ffffff\",\"#f00c0c\"))(100), \n",
    "                 cellwidth = 10,\n",
    "                 cellheight = 10,\n",
    "                 cluster_rows=F,\n",
    "                 cluster_cols=F,\n",
    "                 clustering_distance_cols=\"euclidean\", \n",
    "                 clustering_distance_rows=\"euclidean\",\n",
    "                 clustering_method=\"complete\",\n",
    "                 legend = T,\n",
    "                 treeheight_row=10, treeheight_col=50,\n",
    "                 border_color=F, scale=\"row\", show_colnames=T, show_rownames=T, \n",
    "                 annotation_col=anno, annotation_colors=cols,\n",
    "                 annotation_row = NA,\n",
    "                 annotation_legend = T, \n",
    "                 main='NER genes in foetal and paediatric control and AAI-treated samples')\n",
    "\n",
    "print(heat)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c56184f",
   "metadata": {},
   "source": [
    "# Complex Heatmap package script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06a07a44",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(pheatmap)\n",
    "library(RColorBrewer)\n",
    "library(viridis)\n",
    "library(tidyverse)\n",
    "library(ComplexHeatmap)\n",
    "library(colorRamp2)\n",
    "library(ggplot2)\n",
    "library(colorRamp2)\n",
    "library(AnnotationDbi)\n",
    "library(org.Hs.eg.db)\n",
    "\n",
    "\n",
    "\n",
    "###################################################\n",
    "### combined gene list p53 ######\n",
    "\n",
    "### Set working directory ###\n",
    "\n",
    "setwd('C:/Users/issyl/OneDrive - University of Cambridge/Research Project/DGE analysis AAI/DESeq2 analysis/RNA-Seq-AAI-analysis/normalised counts/Norm counts whole dataset included')\n",
    "\n",
    "### Read in normalised counts data ###\n",
    "read.csv(\"USE_THIS_Normalised_counts_whole_dataset_low_counts_removed_5-4-23.csv\") -> list2\n",
    "\n",
    "##### List1 is the list of preselected GOIs to plot on heatmap, list2 is normalised counts dataset #####\n",
    "setwd('C:/Users/issyl/OneDrive - University of Cambridge/Research Project/DGE analysis AAI/DESeq2 analysis/RNA-Seq-AAI-analysis/normalised counts/Norm counts whole dataset included/gene lists')\n",
    "\n",
    "read.csv(\"combined_gene_list_p53.csv\") -> list1\n",
    "\n",
    "heatmap_data <-merge(x= list1,y= list2, by= \"geneid\", all.x = TRUE)\n",
    "\n",
    "## Make geneid the rownames ##\n",
    "heatmap_data$geneid -> rownames(heatmap_data)\n",
    "head(heatmap_data)\n",
    "\n",
    "### To make just one geneid column ##\n",
    "heatmap_data <- heatmap_data[, -1]\n",
    "head(heatmap_data)\n",
    "\n",
    "### Add gene symbols ###\n",
    "heatmap_data$symbol <-mapIds(org.Hs.eg.db, keys=rownames(heatmap_data), column=\"SYMBOL\", keytype=\"ENSEMBL\")\n",
    "\n",
    "## Make the gene symbols the row names ##\n",
    "heatmap_data$symbol -> rownames(heatmap_data)\n",
    "\n",
    "#### Remove unwanted columns ###\n",
    "## Remove other gene symbol column ##\n",
    "heatmap_data <- heatmap_data[, -25]\n",
    "head(heatmap_data)\n",
    "\n",
    "## Reorder samples to desired order for heatmap ##\n",
    "#### column names####\n",
    "heatmap_data <- heatmap_data[, c(1, 2, 7, 8, 3, 9, 4, 10, 5, 11, 6, 12, 13, 14, 19, 20, 15, 21, 16, 22, 17, 23, 18, 24)]\n",
    "\n",
    "#### Reordering the rownames to show genes in order I want on the heatmap ###\n",
    "heatmap_data <- heatmap_data[c(33, 30, 29, 27, 47, 25, 12, 23, 49, 42, 11, 24, 36, 41, 44, 34, 43, 19,\n",
    "                               13, 1, 4, 14, 15, 17, 48, 45, 32, 7, 52, 31, 37, 28, 18, 3, 21, 22, 50, \n",
    "                               46, 38, 26, 10, 16, 40, 6, 9, 5, 2, 8, 51, 39, 35, 20), ]\n",
    "head(heatmap_data)\n",
    "\n",
    "### Make it a matrix ###\n",
    "as.matrix(heatmap_data) -> heatmap_data\n",
    "\n",
    "#### Need an annotation file too ###\n",
    "setwd(\"C:/Users/issyl/OneDrive - University of Cambridge/Research Project/DGE analysis AAI/DESeq2 analysis/RNA-Seq-AAI-analysis/00_raw data\")\n",
    "read.delim(\"metadataRNAseqAAI2.txt\") -> anno\n",
    "rownames(anno) <- anno$sample\n",
    "anno$sample <- NULL\n",
    "anno[, c(1, 3)] -> anno\n",
    "\n",
    "### Make annotation file order match the counts data order in terms of samples (counts data column names need to match anno row names) ###\n",
    "anno <- anno[c(1, 2, 7, 8, 3, 9, 4, 10, 5, 11, 6, 12, 13, 14, 19, 20, 15, 21, 16, 22, 17, 23, 18, 24),]\n",
    "\n",
    "\n",
    "row_split = rep(\"group1\", 52)\n",
    "row_split[1:17] = \"p53\"\n",
    "row_split[18:29] = \"Apoptosis\"\n",
    "row_split[30:45] = \"NER\"\n",
    "row_split[46:52] = \"Senescence\"\n",
    "\n",
    "column_split = rep(\"group1\", 24)\n",
    "column_split[1:12] = \"foetal\"\n",
    "column_split[13:24] = \"paediatric\"\n",
    "\n",
    "\n",
    "\n",
    "pdf(file=\"Combined Heatmap.pdf\", onefile=TRUE, height=8, width=15, useDingbats = FALSE)\n",
    "ht <- ComplexHeatmap::pheatmap(heatmap_data, name= \"Relative Expression\", color=colorRampPalette(c(\"blue\",\"white\",\"red3\"))(50), \n",
    "                         cellwidth = 10,\n",
    "                         cellheight = 6,\n",
    "                         cluster_rows=F,\n",
    "                         cluster_cols=F,\n",
    "                         clustering_distance_cols=\"euclidean\", \n",
    "                         clustering_distance_rows=\"euclidean\",\n",
    "                         clustering_method=\"complete\",\n",
    "                         legend = T,\n",
    "                         treeheight_row=NA, treeheight_col=NA,\n",
    "                         border_color=F, scale=\"row\", show_colnames=T, show_rownames=T,\n",
    "                         fontsize_row= 7,\n",
    "                         fontsize_col = 8,\n",
    "                         annotation_col=anno, annotation_colors=cols,\n",
    "                         annotation_row = NA,\n",
    "                         annotation_legend = T,\n",
    "                         row_split = row_split,\n",
    "                         column_split = column_split,\n",
    "                         row_title_gp = gpar(fontsize = 8, fontface=\"bold\"),\n",
    "                         main='(H)   Stress response in foetal and paediatric control and AAI-treated samples')\n",
    "\n",
    "print(ht)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17c2d8eb",
   "metadata": {},
   "source": [
    "# Obtaining input files for volcano plots and GO enrichment plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d4140ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### DESeq2 analysis: combining 24hr and 48hr into one 'group' \n",
    "and doing DGE analysis between untreated (0C and 48C) and treated (24+48) ###\n",
    "\n",
    "#### Set working directory ####\n",
    "setwd('C:/Users/issyl/OneDrive - University of Cambridge/Research Project/DGE analysis AAI/DESeq2 analysis/RNA-Seq-AAI-analysis/00_raw data')\n",
    "\n",
    "## Loading counts matrix and metadata\n",
    "\n",
    "cnts <- read.delim(file.choose(), row.names=1)\n",
    "coldata<- read.delim(file.choose(), row.names=1)\n",
    "\n",
    "\n",
    "### Use this for patients study samples\n",
    "coldata <- coldata[,c(\"treatment\", \"timepoint\", \"stage\", \"line\", \"group\")]\n",
    "\n",
    "# Make factors \n",
    "coldata$treatment <- factor(coldata$treatment)\n",
    "coldata$timepoint <- factor(coldata$timepoint)\n",
    "coldata$stage <- factor(coldata$stage)\n",
    "coldata$ID <- factor(coldata$ID)\n",
    "coldata$line <- factor(coldata$line)\n",
    "coldata$group <- factor(coldata$group)\n",
    "\n",
    "\n",
    "### Make a DESeq2 design \n",
    "dds <- DESeqDataSetFromMatrix(countData = cnts,\n",
    "                              colData = coldata,\n",
    "                              design = ~ group)\n",
    "\n",
    "#### Set factor level, have set fetal.control as reference###\n",
    "dds$group <- relevel(dds$group, ref=\"0\")\n",
    "\n",
    "### Filter out feature with very low counts and keep only feature with counts' sum of =>10\n",
    "keep <- rowSums(counts(dds)) >= 10\n",
    "dds <- dds[keep,]\n",
    "\n",
    "#### Run DESeq2 and obtain results table\n",
    "\n",
    "dds <- DESeq(dds)\n",
    "res <- results(dds)\n",
    "head(res)\n",
    "\n",
    "#### Remove version number from Ensembl geneid\n",
    "tmp=gsub(\"\\\\..*\",\"\",row.names(res))\n",
    "rownames(res) <- tmp\n",
    "\n",
    "#Add gene symbol\n",
    "\n",
    "res$symbol <-mapIds(org.Hs.eg.db, keys=rownames(res), column=\"SYMBOL\", keytype=\"ENSEMBL\")\n",
    "head(res)\n",
    "\n",
    "write.csv(res, file= \"full_results_foetal_group_0_24h48h.csv\")\n",
    "\n",
    "#### Use contrast() function to do paed.ctrl vs 24+48h treated and foetal.ctrl vs 24+48h treated ####\n",
    "## Then extract list of DEGs for enrichment analysis SHINY GO ####\n",
    "\n",
    "#### PAED\n",
    "\n",
    "res_paed_ctrl_vs_24h48h <- results(dds, contrast=c(\"group\",\"24h48h\",\"0\"), alpha=0.05)\n",
    "summary(res_paed_ctrl_vs_24h48h)\n",
    "head(res_paed_ctrl_vs_24h48h)\n",
    "\n",
    "\n",
    "tmp=gsub(\"\\\\..*\",\"\",row.names(res_paed_ctrl_vs_24h48h))\n",
    "rownames(res_paed_ctrl_vs_24h48h) <- tmp\n",
    "\n",
    "res_paed_ctrl_vs_24h48h$symbol <-mapIds(org.Hs.eg.db, keys=rownames(res_paed_ctrl_vs_24h48h), column=\"SYMBOL\", keytype=\"ENSEMBL\")\n",
    "\n",
    "#### Obtain just those genes with p-adj <0.05 ###\n",
    "\n",
    "### Reorder and get the list of Differentially Expressed Genes (DEGs) p-values and adjusted p-values\n",
    "\n",
    "resOrdered_paed_ctrl_vs_24h48h <- res_paed_ctrl_vs_24h48h[order(res_paed_ctrl_vs_24h48h$padj),]\n",
    "\n",
    "### Retain DEG with padj < 0.05 and exporting as .csv file\n",
    "resSig_paed_ctrl_vs_24h48h <- subset(resOrdered_paed_ctrl_vs_24h48h, padj < 0.05)\n",
    "head(resSig_paed_ctrl_vs_24h48h)\n",
    "\n",
    "write.csv(as.data.frame(resSig_paed_ctrl_vs_24h48h), file=\"DEGs_paed_ctrl_vs_24h48h.csv\")\n",
    "\n",
    "####Upreg sig DEGs ####\n",
    "sig_DEGs_paed_upreg_24h48h <- subset(resSig_paed_ctrl_vs_24h48h, log2FoldChange >= 0.5)\n",
    "summary(sig_DEGs_paed_upreg_24h48h)\n",
    "\n",
    "write.csv(as.data.frame(sig_DEGs_paed_upreg_24h48h), file=\"sig_upreg_DEGs_paed_24h48h.csv\")\n",
    "\n",
    "#### Downreg sig DEGs ####\n",
    "sig_DEGs_paed_downreg_24h48h <- subset(resSig_paed_ctrl_vs_24h48h, log2FoldChange <= -0.5)\n",
    "summary(sig_DEGs_paed_downreg_24h48h)\n",
    "\n",
    "write.csv(as.data.frame(sig_DEGs_paed_downreg_24h48h), file=\"sig_downreg_DEGs_paed_24h48h.csv\")\n",
    "\n",
    "### Foetal \n",
    "\n",
    "res_foetal_ctrl_vs_24h48h <- results(dds, contrast=c(\"group\",\"24h48h\",\"0\"), alpha=0.05)\n",
    "summary(res_foetal_ctrl_vs_24h48h)\n",
    "tmp=gsub(\"\\\\..*\",\"\",row.names(res_foetal_ctrl_vs_24h48h))\n",
    "rownames(res_foetal_ctrl_vs_24h48h) <- tmp\n",
    "\n",
    "res_foetal_ctrl_vs_24h48h$symbol <-mapIds(org.Hs.eg.db, keys=rownames(res_foetal_ctrl_vs_24h48h), column=\"SYMBOL\", keytype=\"ENSEMBL\")\n",
    "\n",
    "#### Obtain just those genes with p-adj <0.05 ###\n",
    "### Reorder and get the list of Differentially Expressed Genes (DEGs) p-values and adjusted p-values\n",
    "\n",
    "resOrdered_foetal_ctrl_vs_24h48h <- res_foetal_ctrl_vs_24h48h[order(res_foetal_ctrl_vs_24h48h$padj),]\n",
    "\n",
    "### Retain DEG with padj < 0.05 and exporting as .csv file\n",
    "resSig_foetal_ctrl_vs_24h48h <- subset(resOrdered_foetal_ctrl_vs_24h48h, padj < 0.05)\n",
    "head(resSig_foetal_ctrl_vs_24h48h)\n",
    "\n",
    "write.csv(as.data.frame(resSig_foetal_ctrl_vs_24h48h), file=\"DEGs_foetal_ctrl_vs_24h48h.csv\")\n",
    "\n",
    "\n",
    "####Upreg sig DEGs ####\n",
    "sig_DEGs_foetal_upreg_24h48h <- subset(resSig_foetal_ctrl_vs_24h48h, log2FoldChange >= 0.5)\n",
    "summary(sig_DEGs_foetal_upreg_24h48h)\n",
    "\n",
    "write.csv(as.data.frame(sig_DEGs_foetal_upreg_24h48h), file=\"sig_upreg_DEGs_foetal_24h48h.csv\")\n",
    "\n",
    "#### Downreg sig DEGs ####\n",
    "sig_DEGs_foetal_downreg_24h48h <- subset(resSig_foetal_ctrl_vs_24h48h, log2FoldChange <= -0.5)\n",
    "summary(sig_DEGs_foetal_downreg_24h48h)\n",
    "\n",
    "write.csv(as.data.frame(sig_DEGs_foetal_downreg_24h48h), file=\"sig_downreg_DEGs_foetal_24h48h.csv\")\n",
    "\n",
    "\n",
    "########################################\n",
    "\n",
    "#######Volcano plot#####\n",
    "##very basic plot##\n",
    "\n",
    "library(EnhancedVolcano)\n",
    "library(tidyverse)\n",
    "\n",
    "setwd('C:/Users/issyl/OneDrive - University of Cambridge/Research Project/DGE analysis AAI/DESeq2 analysis/RNA-Seq-AAI-analysis/results for volcano plots')\n",
    "\n",
    "read.csv(\"results_foetal_untreated_vs_treated.csv\") -> foetal_res\n",
    "\n",
    "read.csv(\"paed_res_control_vs_24+48+96.csv\") -> paed_res\n",
    "\n",
    "paed_res$X -> row.names(paed_res)\n",
    "paed_res[,-1] -> paed_res\n",
    "\n",
    "foetal_res$X -> row.names(foetal_res)\n",
    "foetal_res[,-1] -> foetal_res\n",
    "\n",
    "\n",
    "#### Paediatric volcano plot ####\n",
    "\n",
    "keyvals <- ifelse(\n",
    "  paed_res$log2FoldChange > 0.5, \"red\", \n",
    "  ifelse(paed_res$log2FoldChange < -0.5, \"royalblue\",\n",
    "         \"black\"))\n",
    "  \n",
    "keyvals[is.na(keyvals)] <- \"black\"\n",
    "names(keyvals)[keyvals == \"red\"] <- \"upregulated\"\n",
    "names(keyvals)[keyvals == \"royalblue\"] <- \"downregulated\"\n",
    "\n",
    "\n",
    "\n",
    "paed_vplot <- EnhancedVolcano(paed_res, lab=paed_res$symbol, x=\"log2FoldChange\", y=\"padj\", xlim = c(-6, 10), ylim = c(0, 25), title= \"Paediatric Untreated vs AAI-Treated DEGs\", subtitle= NULL, pCutoff= 10e-2, FCcutoff= 0.5, labSize= 3.0, \n",
    "                colCustom = keyvals, colAlpha=0.5, cutoffLineType = \"dashed\", cutoffLineCol = \"black\", cutoffLineWidth = 0.8, \n",
    "                legendPosition = \"right\", legendLabSize = 12, legendIconSize = 4, selectLab = c(\"MDM2\", \"BAX\", \"DDB2\", \"CCND1\", \"GADD45A\", \"CDKN1A\", \"XPC\", \"WWTR1\", \"FAS\", \"TP53I3\"), gridlines.major=TRUE,\n",
    "                gridlines.minor=FALSE, caption= NULL, titleLabSize= 12, subtitleLabSize= 10, ylab= \"-log10(padj)\", axisLabSize=8)\n",
    "                                                                                                                      \n",
    "\n",
    "\n",
    "\n",
    "paed_vplot + ggplot2:: coord_cartesian(xlim=c(-6, 10)) +\n",
    "                                         ggplot2::scale_x_continuous(\n",
    "                                           breaks=seq(-6,10, 2))\n",
    "                                         \n",
    "                                  \n",
    "\n",
    "#### Foetal volcano plot ####\n",
    "\n",
    "EnhancedVolcano(foetal_res, lab=foetal_res$symbol, x=\"log2FoldChange\", y=\"padj\", xlim= c(-4.75, 6), ylim = c(0, 1), title= \"Foetal Untreated vs AAI-Treated DEGs\", pCutoff= 10e-2, FCcutoff= 1, labSize= 5.0, \n",
    "                col=c(\"black\", \"green4\", \"black\", \"red3\", colAlpha=0.8), cutoffLineType = \"dashed\", cutoffLineCol = \"black\", cutoffLineWidth = 0.8, \n",
    "                legendPosition = \"right\", legendLabSize = 10, legendIconSize = 5)\n",
    "\n",
    "######################################################################################\n",
    "#### Enrichment Plots #######\n",
    "\n",
    "###### Plots to illustrate functional enrichment analysis of top differentially expressed genes ####\n",
    "\n",
    "library(tidyverse)\n",
    "library(ggplot2)\n",
    "\n",
    "setwd('C:/Users/issyl/OneDrive - University of Cambridge/Research Project/DGE analysis AAI/DESeq2 analysis/RNA-Seq-AAI-analysis/sig DEG lists for enrichment plots/enrichment plots')\n",
    "read.csv(\"metadata_paed_enrichment_plot.csv\", header = TRUE) -> enrichment_paed\n",
    "\n",
    "as.factor(enrichment_paed$pathway) -> enrichment_paed$pathway \n",
    "as.factor(enrichment_paed$fold.change) -> enrichment_paed$fold.change\n",
    "\n",
    "enrichment_paed$fold.change <- factor(enrichment_paed$fold.change, levels=c(\"Top 10 downregulated\", \"Top 10 upregulated\"))\n",
    "\n",
    "enrichment_paed$pathway <- factor(enrichment_paed$pathway, levels=unique(enrichment_paed$pathway))\n",
    "\n",
    "\n",
    "ggplot(enrichment_paed, aes(fill=fold.change)) + geom_bar(stat=\"identity\", alpha=0.6, aes(y=pathway, x=fold.enrichment)) +\n",
    "  labs(x= \"Fold Enrichment\", y= NULL, colour= NULL) +ggtitle(\"Biological Processes Enrichment in Paediatric Untreated compared to AAI-Treated Organoids\") +\n",
    "  theme(plot.title=element_text(size=12)) +\n",
    "  theme(plot.title=element_text(hjust=0.8)) + theme(legend.title=element_blank()) +\n",
    "  scale_fill_manual(values= c(\"Top 10 upregulated\" = \"red2\", \"Top 10 downregulated\" = \"royalblue\")) +\n",
    "  theme(panel.grid.major = element_blank()) + theme(panel.grid.minor = element_blank()) +\n",
    "  scale_x_continuous(limits= c(0, 10), breaks= seq(0, 10, by= 4)) + \n",
    "  theme(plot.title = element_text(margin = margin(b=20))) +\n",
    "  theme(axis.text.y = element_text(size=11)) +\n",
    "  theme(legend.text = element_text(size=10)) +\n",
    "  theme(plot.title = element_text(face = \"bold\"))\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74117677",
   "metadata": {},
   "source": [
    "# Plotting top GO enrichment hits, as a bar chart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "778948e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "######################################################################################\n",
    "#### Enrichment Plots #######\n",
    "\n",
    "###### Plots to illustrate functional enrichment analysis of top differentially expressed genes ####\n",
    "\n",
    "library(tidyverse)\n",
    "library(ggplot2)\n",
    "\n",
    "setwd('C:/Users/issyl/OneDrive - University of Cambridge/Research Project/DGE analysis AAI/DESeq2 analysis/RNA-Seq-AAI-analysis/sig DEG lists for enrichment plots/enrichment plots')\n",
    "read.csv(\"metadata_paed_enrichment_plot.csv\", header = TRUE) -> enrichment_paed\n",
    "\n",
    "as.factor(enrichment_paed$pathway) -> enrichment_paed$pathway \n",
    "as.factor(enrichment_paed$fold.change) -> enrichment_paed$fold.change\n",
    "\n",
    "enrichment_paed$fold.change <- factor(enrichment_paed$fold.change, levels=c(\"Top 10 downregulated\", \"Top 10 upregulated\"))\n",
    "\n",
    "enrichment_paed$pathway <- factor(enrichment_paed$pathway, levels=unique(enrichment_paed$pathway))\n",
    "\n",
    "\n",
    "ggplot(enrichment_paed, aes(fill=fold.change)) + geom_bar(stat=\"identity\", alpha=0.6, aes(y=pathway, x=fold.enrichment)) +\n",
    "  labs(x= \"Fold Enrichment\", y= NULL, colour= NULL) +ggtitle(\"Biological Processes Enrichment in Paediatric Untreated compared to AAI-Treated Organoids\") +\n",
    "  theme(plot.title=element_text(size=12)) +\n",
    "  theme(plot.title=element_text(hjust=0.8)) + theme(legend.title=element_blank()) +\n",
    "  scale_fill_manual(values= c(\"Top 10 upregulated\" = \"red2\", \"Top 10 downregulated\" = \"royalblue\")) +\n",
    "  theme(panel.grid.major = element_blank()) + theme(panel.grid.minor = element_blank()) +\n",
    "  scale_x_continuous(limits= c(0, 10), breaks= seq(0, 10, by= 4)) + \n",
    "  theme(plot.title = element_text(margin = margin(b=20))) +\n",
    "  theme(axis.text.y = element_text(size=11)) +\n",
    "  theme(legend.text = element_text(size=10)) +\n",
    "  theme(plot.title = element_text(face = \"bold\"))\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R 4.2.2",
   "language": "R",
   "name": "ir422"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
